---
title: "Gene Level Homology Julin Ask"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{bash, eval = F}
# blast/2.12.0+
blastp \
 -db ../Sequences/${Seq_1} \
 -query ../Sequences/${Seq_2} \
 -max_target_seqs 5 \
 -out ${Name_1}_vs_${Name_2}.out \
 -evalue 1e-5 \
 -num_threads 4 \
 -outfmt "6 std qlen slen"

blastp \
 -db ../Sequences/${Seq_2} \
 -query ../Sequences/${Seq_1} \
 -max_target_seqs 5 \
 -out ${Name_2}_vs_${Name_1}.out \
 -evalue 1e-5 \
 -num_threads 4 \
 -outfmt "6 std qlen slen"
```

### Comparing the ancestral B.rapa and B.oleracea to DaAe

```{r}
library(tidyverse)
headers <- c("query.acc.ver","subject.acc.ver","pct.identity","alignment.length","mismatches","gap.opens","q.start","q.end","s.start","s.end","evalue","bit.score","query.length","subject.length")
```
## DaAE
```{r}
# Finding the RBH between each genome/subgenome
# Find the RBH between B.rapa and B.oleracea
Brapa_vs_Boleracea <- read_table("BlastP/Brapa_Z1_V2_vs_Boleracea_HDEM.out.gz", col_names = headers)
Boleracea_vs_Brapa <- read_table("BlastP/Boleracea_HDEM_vs_Brapa_Z1_V2.out.gz", col_names = headers)

# Filter for best hits
Brapa_vs_Boleracea <- Brapa_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Boleracea_vs_Brapa <- Boleracea_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Boleracea <- inner_join(Brapa_vs_Boleracea,
                                  Boleracea_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for DaAe_A and Brapa
# Find the RBH between DaAe_A and Brapa
Brapa_vs_DaAe <- read_table("BlastP/Brapa_Z1_V2_vs_DaAe.out.gz", col_names = headers) %>% filter(!grepl("_Random_", query.acc.ver))
Brapa_vs_DaAe_A <- Brapa_vs_DaAe %>%
  filter(grepl("BnachrA",query.acc.ver))
DaAe_vs_Brapa <- read_table("BlastP/DaAe_vs_Brapa_Z1_V2.out.gz", col_names = headers) %>% filter(!grepl("_Random_", subject.acc.ver))
DaAe_A_vs_Brapa <- DaAe_vs_Brapa %>%
  filter(grepl("BnachrA",subject.acc.ver))

# Filter for best hits
Brapa_vs_DaAe_A <- Brapa_vs_DaAe_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
DaAe_A_vs_Brapa <- DaAe_A_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_DaAe_A <- inner_join(Brapa_vs_DaAe_A,
                                  DaAe_A_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for DaAe_C and Brapa
# Find the RBH between DaAe_C and Brapa
Brapa_vs_DaAe_C <- Brapa_vs_DaAe %>%
  filter(grepl("BnachrC",query.acc.ver))
DaAe_C_vs_Brapa <- DaAe_vs_Brapa %>%
  filter(grepl("BnachrC",subject.acc.ver))

# Filter for best hits
Brapa_vs_DaAe_C <- Brapa_vs_DaAe_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
DaAe_C_vs_Brapa <- DaAe_C_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_DaAe_C <- inner_join(Brapa_vs_DaAe_C,
                                  DaAe_C_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for DaAe_A and Boleracea
# Find the RBH between DaAe_A and Boleracea
Boleracea_vs_DaAe <- read_table("BlastP/Boleracea_HDEM_vs_DaAe.out.gz", col_names = headers) %>% filter(!grepl("_Random_", query.acc.ver))
Boleracea_vs_DaAe_A <- Boleracea_vs_DaAe %>%
  filter(grepl("BnachrA",query.acc.ver))
DaAe_vs_Boleracea <- read_table("BlastP/DaAe_vs_Boleracea_HDEM.out.gz", col_names = headers) %>% filter(!grepl("_Random_", subject.acc.ver))
DaAe_A_vs_Boleracea <- DaAe_vs_Boleracea %>%
  filter(grepl("BnachrA",subject.acc.ver))

# Filter for best hits
Boleracea_vs_DaAe_A <- Boleracea_vs_DaAe_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
DaAe_A_vs_Boleracea <- DaAe_A_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_DaAe_A <- inner_join(Boleracea_vs_DaAe_A,
                                  DaAe_A_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for DaAe_C and Boleracea
# Find the RBH between DaAe_C and Boleracea
Boleracea_vs_DaAe_C <- Boleracea_vs_DaAe %>%
  filter(grepl("BnachrC",query.acc.ver))
DaAe_C_vs_Boleracea <- DaAe_vs_Boleracea %>%
  filter(grepl("BnachrC",subject.acc.ver))

# Filter for best hits
Boleracea_vs_DaAe_C <- Boleracea_vs_DaAe_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
DaAe_C_vs_Boleracea <- DaAe_C_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_DaAe_C <- inner_join(Boleracea_vs_DaAe_C,
                                  DaAe_C_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for DaAe_A and DaAe_C
# Find the RBH between DaAe_A and DaAe_C
DaAe_A_vs_DaAe_C <- read_table("BlastP/DaAe_A_vs_DaAe_C.out.gz", col_names = headers) %>% filter(!grepl("_Random_", query.acc.ver))
DaAe_C_vs_DaAe_A <- read_table("BlastP/DaAe_C_vs_DaAe_A.out.gz", col_names = headers) %>% filter(!grepl("_Random_", subject.acc.ver))

# Filter for best hits
DaAe_A_vs_DaAe_C <- DaAe_A_vs_DaAe_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
DaAe_C_vs_DaAe_A <- DaAe_C_vs_DaAe_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
DaAe_A_and_DaAe_C <- inner_join(DaAe_A_vs_DaAe_C,
                                  DaAe_C_vs_DaAe_A,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))
```

For simplicity of the analysis, orthogonal genes between B.rapa and B.oleracea were used to filter potential sites of homeologous exchange between DaAe's two subgenomes, A and C. Homeologous exchange is the exchange of genetic material from one subgenome to the other. In this case it would be the conversion of an A subgenome gene to a C subgenome gene or vice versa. Brassica napus is an allo-tetraploid containing two diploid subgenomes A and C, meaning homeologous exchange can result in homeologous gene ratios of 2:2, 3:1, and 4:0 corresponding to a reciprocal, partial, or complete gene transversion. Because the present assembly is unphased, attempting to identify potential 3:1 homeologous gene ratios is inhibited by the assembler program either selecting one of the two homeologous genes or creating a mashup of the two genes when creating the consensus sequence. In both cases, the true underlying biology is not being accurately represented in the assembly sequence. For this reason, complete gene transversions were the group of focus in this homeologous analysis.

## Find potential sites where a C gene in the C subgenome was converted to an A gene

For a potential homeologous gene pair to be considered as a possible site of homeologous exchange, two requirements must be met. First, the gene must align better to its homeolog than it does to its ortholog. Second, the gene must also align better to its homeolog's ortholog than it does to its ortholog. If an annotation is incomplete or erroneous, it can create both false positive and false negative results.

Compare alignments
If HE both true:
  Rule 1: An-Cn > Co-Cn
  Rule 2: Ar-Cn > Co-Cn
  
```{r}
HE_of_A <- function(ArCo, ArAn, ArCn, CoCn, AnCn, name){
  # Filter
  Filtered_ArAn <- ArAn %>%
    filter(subject.acc.ver %in% ArCo$subject.acc.ver)
  Filtered_AnCn <- AnCn %>%
    filter(subject.acc.ver %in% Filtered_ArAn$query.acc.ver)

  # Smaller data frames
  Small_AnCn <- Filtered_AnCn %>%
    mutate(pct.identity = (pct.identity.x + pct.identity.y)/2,
           alignment.length = (alignment.length.x + alignment.length.y)/2,
           bit.score = (bit.score.x + bit.score.y) / 2,) %>%
    select(query.acc.ver, subject.acc.ver,
           pct.identity,
           alignment.length,
           bit.score)
  Small_CoCn <- CoCn %>%
    mutate(pct.identity = (pct.identity.x + pct.identity.y)/2,
           alignment.length = (alignment.length.x + alignment.length.y)/2,
           bit.score = (bit.score.x + bit.score.y) / 2,) %>%
    select(query.acc.ver, subject.acc.ver,
           pct.identity,
           alignment.length,
           bit.score)
  Small_ArCn <- ArCn %>%
    mutate(pct.identity = (pct.identity.x + pct.identity.y)/2,
           alignment.length = (alignment.length.x + alignment.length.y)/2,
           bit.score = (bit.score.x + bit.score.y) / 2,) %>%
    select(query.acc.ver, subject.acc.ver,
           pct.identity,
           alignment.length,
           bit.score)
  
  # Rule 1 genes An-Cn > Co-Cn
  Passed_Rule_1 <- Small_AnCn %>%
    left_join(Small_CoCn, by = c("query.acc.ver" = "query.acc.ver"), suffix = c(paste0("_", name) ,"_Oleracea")) %>%
    filter(get(paste0("bit.score_", name)) > bit.score_Oleracea)
  
  # Rule 2 genes Ar-Cn > Co-Cn
  Passed_Rule_2 <- Small_ArCn %>%
    left_join(Small_CoCn, by = c("query.acc.ver" = "query.acc.ver"), suffix = c("_Rapa","_Oleracea")) %>%
    filter(bit.score_Rapa > bit.score_Oleracea)
  
  # Combine rules
  Small_Rule_1 <- Passed_Rule_1 %>%
    select(query.acc.ver, subject.acc.ver = paste0("subject.acc.ver_", name))
  Small_Rule_2 <- Passed_Rule_2 %>%
    select(query.acc.ver, subject.acc.ver = subject.acc.ver_Rapa)
  Combined_Rules <- Small_Rule_1 %>%
    filter(query.acc.ver %in% Small_Rule_2$query.acc.ver)
  Possible_Double_A <- Combined_Rules %>%
    left_join(Filtered_AnCn)
  
  return(Possible_Double_A)
}

DaAe_Double_A <- HE_of_A(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_DaAe_A,
                         ArCn = Brapa_and_DaAe_C,
                         CoCn = Boleracea_and_DaAe_C,
                         AnCn = DaAe_A_and_DaAe_C,
                         name = "DaAe")
```

## Find potential sites where an A gene in the A subgenome was converted to a C gene


Compare alignments
If HE both true:
  Rule 1: An-Cn > Ar-An
  Rule 2: Co-An > Ar-An

```{r}
HE_of_C <- function(ArCo, ArAn, CoAn, CoCn, AnCn, name){
  # Filter
  Filtered_CoCn <- CoCn %>%
    filter(subject.acc.ver %in% ArCo$query.acc.ver)
  Filtered_AnCn <- AnCn %>%
    filter(query.acc.ver %in% Filtered_CoCn$query.acc.ver)

  # Smaller data frames
  Small_AnCn <- Filtered_AnCn %>%
    mutate(pct.identity = (pct.identity.x + pct.identity.y)/2,
           alignment.length = (alignment.length.x + alignment.length.y)/2,
           bit.score = (bit.score.x + bit.score.y) / 2,) %>%
    select(query.acc.ver, subject.acc.ver,
           pct.identity,
           alignment.length,
           bit.score)
  Small_ArAn <- ArAn %>%
    mutate(pct.identity = (pct.identity.x + pct.identity.y)/2,
           alignment.length = (alignment.length.x + alignment.length.y)/2,
           bit.score = (bit.score.x + bit.score.y) / 2,) %>%
    select(query.acc.ver, subject.acc.ver,
           pct.identity,
           alignment.length,
           bit.score)
  Small_CoAn <- CoAn %>%
    mutate(pct.identity = (pct.identity.x + pct.identity.y)/2,
           alignment.length = (alignment.length.x + alignment.length.y)/2,
           bit.score = (bit.score.x + bit.score.y) / 2,) %>%
    select(query.acc.ver, subject.acc.ver,
           pct.identity,
           alignment.length,
           bit.score)
  
  # Rule 1 genes An-Cn > Co-Cn
  Passed_Rule_1 <- Small_AnCn %>%
    left_join(Small_ArAn, by = c("subject.acc.ver" = "query.acc.ver"), suffix = c(paste0("_", name) ,"_Rapa")) %>%
    filter(get(paste0("bit.score_", name)) > bit.score_Rapa)
  
  # Rule 2 genes Ar-Cn > Co-Cn
  Passed_Rule_2 <- Small_CoAn %>%
    left_join(Small_ArAn, by = c("query.acc.ver" = "query.acc.ver"), suffix = c("_Oleracea","_Rapa")) %>%
    filter(bit.score_Oleracea > bit.score_Rapa)
  
  # Combine rules
  Small_Rule_1 <- Passed_Rule_1 %>%
    select(query.acc.ver, subject.acc.ver)
  Small_Rule_2 <- Passed_Rule_2 %>%
    select(query.acc.ver, subject.acc.ver = subject.acc.ver_Oleracea)
  Combined_Rules <- Small_Rule_1 %>%
    filter(subject.acc.ver %in% Small_Rule_2$query.acc.ver)
  Possible_Double_C <- Combined_Rules %>%
    left_join(Filtered_AnCn)
  
  return(Possible_Double_C)
}

DaAe_Double_C <- HE_of_C(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_DaAe_A,
                         CoAn = Boleracea_and_DaAe_A,
                         CoCn = Boleracea_and_DaAe_C,
                         AnCn = DaAe_A_and_DaAe_C,
                         name = "DaAe")
```


## Looking at coverage of genes in BED files

To look at the coverage of the genes listed in the BED files, the trimmed 10X reads were mapped to a combined progenitor genome containing A01-A10 from B.rapa and C01-C09 from B.oleracea using BWA. To look at possible sites of homeologous exchange, a subsample of reads with unique trustworthy mapping locations was desired. These were reads which could realiably be described as coming from either the A or C subgenome. Alignments were filtered to only contain reads which had a MAPQ of 5 or greater, were properly paired, had no supplementary alignments, and were primary alignments. This subsample of reads was then mapped to the DaAe assembly and filtered for alignments with a MAPQ of 5 or greater. The two alignment files, filtered progenitor and DaAe, were then used to calculate the coverage of their potential homeologous genes using bedtools coverage and the previously created BED files.

```{r}
## Length of gene was also factored into coverage estimates
## For simplicity, the coverage of all genes was calculated then subset for the potential HE genes
## Starting with the coverage of the progenitor genes

## Read in the calculated coverages
Coverage_Readin <- function(filename){
  df <- read_delim(filename,
                   col_names = T,
                   delim = "\t") %>%
    filter(!is.na(Chromosome)) %>%
    mutate(Adjusted_Coverage = round(Coverage / Size, 2),
           Position = floor((End + Start) / 2))
  return(df)
}


Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Gene_Coverage.tsv.gz")
DaAe_Coverage <- Coverage_Readin("Combined_Coverages/Combined_DaAe_Gene_Coverage.tsv.gz")
```

# Double A in Progenitor genes

```{r}
DA_Pseudo_napus_Formatter <- function(ArCo,
                                      Pseudo_napus_Coverage,
                                      ArAn,
                                      Double_A,
                                      cutoff = 2.5){
  df <- ArAn %>%
    filter(query.acc.ver %in% Double_A$subject.acc.ver) %>%
    select(subject.acc.ver) %>%
    inner_join(ArCo, by = "subject.acc.ver") %>%
    select(query.acc.ver, subject.acc.ver) %>%
    mutate(subject.acc.ver = str_remove(subject.acc.ver, "Parent=")) %>%
    left_join(Pseudo_napus_Coverage %>%
                select(GeneId, Adjusted_Coverage), by = c("subject.acc.ver" = "GeneId")) %>%
    select(Rapa = subject.acc.ver, Oleracea = query.acc.ver, Rapa_coverage = Adjusted_Coverage) %>%
    left_join(Pseudo_napus_Coverage %>%
                select(GeneId, Adjusted_Coverage), by = c("Oleracea" = "GeneId")) %>%
    select(Rapa, Oleracea, Rapa_coverage, Oleracea_coverage = Adjusted_Coverage) %>%
    filter(Rapa_coverage > 0 & Rapa_coverage > cutoff*Oleracea_coverage) %>%
    mutate(Ratio = round(Rapa_coverage/Oleracea_coverage,2)) %>%
    arrange(Rapa)
  return(df)
}

DA_Pseudo_napus_DaAe <- DA_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                      Pseudo_napus_Coverage,
                                      Brapa_and_DaAe_A,
                                      DaAe_Double_A,
                                      cutoff = 2.5)

```

# Double C in Progenitor genes

```{r}
DC_Pseudo_napus_Formatter <- function(ArCo,
                                    Pseudo_napus_Coverage,
                                    CoCn,
                                    Double_C,
                                    cutoff = 2.5){
  df <- CoCn %>%
    filter(query.acc.ver %in% Double_C$query.acc.ver) %>%
    select(subject.acc.ver) %>%
    inner_join(ArCo, by = c("subject.acc.ver" = "query.acc.ver")) %>%
    select(qseqid = subject.acc.ver.y, sseqid = subject.acc.ver) %>%
    mutate(qseqid = str_remove(qseqid, "Parent=")) %>%
    left_join(Pseudo_napus_Coverage %>%
                select(GeneId, Adjusted_Coverage), by = c("qseqid" = "GeneId")) %>%
    select(Rapa = qseqid, Oleracea = sseqid, Rapa_coverage = Adjusted_Coverage) %>%
    left_join(Pseudo_napus_Coverage %>%
                select(GeneId, Adjusted_Coverage), by = c("Oleracea" = "GeneId")) %>%
    select(Rapa, Oleracea, Rapa_coverage, Oleracea_coverage = Adjusted_Coverage) %>%
    filter(Oleracea_coverage > 0 & Oleracea_coverage > cutoff*Rapa_coverage) %>%
    mutate(Ratio = round(Oleracea_coverage/Rapa_coverage,2)) %>%
    arrange(Rapa)
  return(df)
}

DC_Pseudo_napus_DaAe <- DC_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                                  Pseudo_napus_Coverage,
                                                  Boleracea_and_DaAe_C,
                                                  DaAe_Double_C,
                                                  cutoff = 2.5)

```

# Double A and Double C in Napus genes

```{r}
Double_Pseudo_napus_Formatter <- function(Query_Coverage, Double, cutoff_min = 0.5, cutoff_max = 1.5){
  df <- Double %>%
    select(qseqid = query.acc.ver, sseqid = subject.acc.ver) %>%
    left_join(Query_Coverage, by = c("sseqid" = "GeneId")) %>%
    select(Genome_A = sseqid, Genome_C = qseqid, Genome_A_coverage = Adjusted_Coverage) %>%
    left_join(Query_Coverage, by = c("Genome_C" = "GeneId")) %>%
    select(Genome_A, Genome_C, Genome_A_coverage, Genome_C_coverage = Adjusted_Coverage) %>%
    mutate(Ratio = round(Genome_A_coverage / Genome_C_coverage, 2)) %>%
    filter(Genome_A_coverage > 0 & Genome_C_coverage > 0 & Ratio >= cutoff_min & Ratio <= cutoff_max) %>%
    arrange(Genome_A)
  return(df)
}

DA_DaAe <- Double_Pseudo_napus_Formatter(DaAe_Coverage, DaAe_Double_A)
DC_DaAe <- Double_Pseudo_napus_Formatter(DaAe_Coverage, DaAe_Double_C)
```

## Combining the progenitor and napus coverages

```{r}

Coverage_Combiner <- function(ArCo, ArCo_Coverage, Query_Alignment, Query_Alignment_2, Query_Coverage){
  df <- Query_Alignment %>%
    mutate(subject.acc.ver = str_remove(subject.acc.ver, "Parent=")) %>%
    filter(subject.acc.ver %in% ArCo_Coverage$Rapa) %>%
    filter(query.acc.ver %in% Query_Coverage$Genome_A) %>%
    select(Rapa = subject.acc.ver, Genome_A = query.acc.ver) %>%
    left_join(ArCo %>% mutate(subject.acc.ver = str_remove(subject.acc.ver, "Parent=")),
              by = c("Rapa" = "subject.acc.ver")) %>%
    filter(query.acc.ver %in% ArCo_Coverage$Oleracea) %>%
    select(Rapa, Oleracea = query.acc.ver, Genome_A) %>%
    left_join(Query_Alignment_2, by = c("Oleracea" = "subject.acc.ver")) %>%
    filter(query.acc.ver %in% Query_Coverage$Genome_C) %>%
    select(Rapa, Oleracea, Genome_A, Genome_C = "query.acc.ver") %>%
    left_join(Query_Coverage %>%
                select(Genome_A, Ratio)) %>%
    select(Rapa, Oleracea, Genome_A, Genome_C, Napus_Ratio = Ratio) %>%
    left_join(ArCo_Coverage %>%
                select(Rapa, Ratio)) %>%
    select(Rapa, Oleracea, Genome_A, Genome_C, Napus_Ratio, ArCo_Ratio = Ratio) %>%
    arrange(Rapa) %>%
    mutate(ArCo_Ratio = ifelse(is.infinite(ArCo_Ratio),
                                     5,
                                     ifelse(ArCo_Ratio > 5,
                                            5,
                                            ArCo_Ratio)),
           Group = 1:nrow(.))
  return(df)
}

DA_Combined_DaAe <- Coverage_Combiner(Brapa_and_Boleracea, DA_Pseudo_napus_DaAe, Brapa_and_DaAe_A, Boleracea_and_DaAe_C, DA_DaAe)
DC_Combined_DaAe <- Coverage_Combiner(Brapa_and_Boleracea, DC_Pseudo_napus_DaAe, Brapa_and_DaAe_A, Boleracea_and_DaAe_C, DC_DaAe)

```

```{r}
save(DA_Combined_DaAe, DC_Combined_DaAe, file = "Julin_DaAe_gene_doublings.RData")
```

### Extracting positions of genes

```{r}

Position_Extract <- function(Combined_Coverage, Pseudo_napus_Coverage, Napus_Coverage){
  GeneId <- c(Combined_Coverage$Rapa,Combined_Coverage$Oleracea,Combined_Coverage$Genome_A,Combined_Coverage$Genome_C)
  Group = rep(Combined_Coverage$Group,4)
  Assembly = rep(c("Pseudo_napus", "Napus"), each = length(GeneId)/2)
  df <- data.frame(GeneId, Group, Assembly, stringsAsFactors = F)
  df <- df %>%
    left_join(rbind(Pseudo_napus_Coverage, Napus_Coverage) %>%
                select(GeneId, Chromosome, Position))
  return(df)
}

DA_DaAe_Positions <- Position_Extract(DA_Combined_DaAe, Pseudo_napus_Coverage, DaAe_Coverage)
DC_DaAe_Positions <- Position_Extract(DC_Combined_DaAe, Pseudo_napus_Coverage, DaAe_Coverage)

save(DA_DaAe_Positions, DC_DaAe_Positions, file = "Julin_DaAe_Gene_Positions.RData")

rm(list = ls(pattern = "*DaAe*"))
```

## ZS11
```{r}
# Repeat for DaAe_A and Brapa
# Find the RBH between DaAe_A and Brapa
Brapa_vs_ZS11 <- read_table("BlastP/Brapa_Z1_V2_vs_ZS11.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Brapa_vs_ZS11_A <- Brapa_vs_ZS11 %>%
  filter(grepl("BnaA",query.acc.ver))
ZS11_vs_Brapa <- read_table("BlastP/ZS11_vs_Brapa_Z1_V2.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
ZS11_A_vs_Brapa <- ZS11_vs_Brapa %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Brapa_vs_ZS11_A <- Brapa_vs_ZS11_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
ZS11_A_vs_Brapa <- ZS11_A_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_ZS11_A <- inner_join(Brapa_vs_ZS11_A,
                                  ZS11_A_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for ZS11_C and Brapa
# Find the RBH between ZS11_C and Brapa
Brapa_vs_ZS11_C <- Brapa_vs_ZS11 %>%
  filter(grepl("BnaC",query.acc.ver))
ZS11_C_vs_Brapa <- ZS11_vs_Brapa %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Brapa_vs_ZS11_C <- Brapa_vs_ZS11_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
ZS11_C_vs_Brapa <- ZS11_C_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_ZS11_C <- inner_join(Brapa_vs_ZS11_C,
                                  ZS11_C_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for ZS11_A and Boleracea
# Find the RBH between ZS11_A and Boleracea
Boleracea_vs_ZS11 <- read_table("BlastP/Boleracea_HDEM_vs_ZS11.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Boleracea_vs_ZS11_A <- Boleracea_vs_ZS11 %>%
  filter(grepl("BnaA",query.acc.ver))
ZS11_vs_Boleracea <- read_table("BlastP/ZS11_vs_Boleracea_HDEM.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
ZS11_A_vs_Boleracea <- ZS11_vs_Boleracea %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Boleracea_vs_ZS11_A <- Boleracea_vs_ZS11_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
ZS11_A_vs_Boleracea <- ZS11_A_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_ZS11_A <- inner_join(Boleracea_vs_ZS11_A,
                                  ZS11_A_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for ZS11_C and Boleracea
# Find the RBH between ZS11_C and Boleracea
Boleracea_vs_ZS11_C <- Boleracea_vs_ZS11 %>%
  filter(grepl("BnaC",query.acc.ver))
ZS11_C_vs_Boleracea <- ZS11_vs_Boleracea %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Boleracea_vs_ZS11_C <- Boleracea_vs_ZS11_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
ZS11_C_vs_Boleracea <- ZS11_C_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_ZS11_C <- inner_join(Boleracea_vs_ZS11_C,
                                  ZS11_C_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for ZS11_A and ZS11_C
# Find the RBH between ZS11_A and ZS11_C
ZS11_A_vs_ZS11_C <- read_table("BlastP/ZS11_A_vs_ZS11_C.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
ZS11_C_vs_ZS11_A <- read_table("BlastP/ZS11_C_vs_ZS11_A.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))

# Filter for best hits
ZS11_A_vs_ZS11_C <- ZS11_A_vs_ZS11_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
ZS11_C_vs_ZS11_A <- ZS11_C_vs_ZS11_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
ZS11_A_and_ZS11_C <- inner_join(ZS11_A_vs_ZS11_C,
                                  ZS11_C_vs_ZS11_A,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

ZS11_Double_A <- HE_of_A(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_ZS11_A,
                         ArCn = Brapa_and_ZS11_C,
                         CoCn = Boleracea_and_ZS11_C,
                         AnCn = ZS11_A_and_ZS11_C,
                         name = "ZS11")


ZS11_Double_C <- HE_of_C(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_ZS11_A,
                         CoAn = Boleracea_and_ZS11_A,
                         CoCn = Boleracea_and_ZS11_C,
                         AnCn = ZS11_A_and_ZS11_C,
                         name = "ZS11")

Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Gene_Coverage.tsv.gz")
ZS11_Coverage <- Coverage_Readin("Combined_Coverages/Combined_ZS11_Gene_Coverage.tsv.gz")

DA_Pseudo_napus_ZS11 <- DA_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                      Pseudo_napus_Coverage,
                                      Brapa_and_ZS11_A,
                                      ZS11_Double_A,
                                      cutoff = 2.5)


DC_Pseudo_napus_ZS11 <- DC_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                                  Pseudo_napus_Coverage,
                                                  Boleracea_and_ZS11_C,
                                                  ZS11_Double_C,
                                                  cutoff = 2.5)


DA_ZS11 <- Double_Pseudo_napus_Formatter(ZS11_Coverage, ZS11_Double_A)
DC_ZS11 <- Double_Pseudo_napus_Formatter(ZS11_Coverage, ZS11_Double_C)

DA_Combined_ZS11 <- Coverage_Combiner(Brapa_and_Boleracea, DA_Pseudo_napus_ZS11, Brapa_and_ZS11_A, Boleracea_and_ZS11_C, DA_ZS11)
DC_Combined_ZS11 <- Coverage_Combiner(Brapa_and_Boleracea, DC_Pseudo_napus_ZS11, Brapa_and_ZS11_A, Boleracea_and_ZS11_C, DC_ZS11)

save(DA_Combined_ZS11, DC_Combined_ZS11, file = "Julin_ZS11_gene_doublings.RData")

DA_ZS11_Positions <- Position_Extract(DA_Combined_ZS11, Pseudo_napus_Coverage, ZS11_Coverage)
DC_ZS11_Positions <- Position_Extract(DC_Combined_ZS11, Pseudo_napus_Coverage, ZS11_Coverage)

save(DA_ZS11_Positions, DC_ZS11_Positions, file = "Julin_ZS11_Gene_Positions.RData")

rm(list = ls(pattern = "*ZS11*"))
```

## Darmor_Bzh_V10
```{r}
# Repeat for DaAe_A and Brapa
# Find the RBH between DaAe_A and Brapa
Brapa_vs_Darmor_Bzh_V10 <- read_table("BlastP/Brapa_Z1_V2_vs_Darmor-Bzh_V10.out.gz", col_names = headers) %>% filter(!grepl("scaffold", query.acc.ver))
Brapa_vs_Darmor_Bzh_V10_A <- Brapa_vs_Darmor_Bzh_V10 %>%
  filter(grepl("^A",query.acc.ver))
Darmor_Bzh_V10_vs_Brapa <- read_table("BlastP/Darmor-Bzh_V10_vs_Brapa_Z1_V2.out.gz", col_names = headers) %>% filter(!grepl("scaffold", subject.acc.ver))
Darmor_Bzh_V10_A_vs_Brapa <- Darmor_Bzh_V10_vs_Brapa %>%
  filter(grepl("^A",subject.acc.ver))

# Filter for best hits
Brapa_vs_Darmor_Bzh_V10_A <- Brapa_vs_Darmor_Bzh_V10_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Darmor_Bzh_V10_A_vs_Brapa <- Darmor_Bzh_V10_A_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Darmor_Bzh_V10_A <- inner_join(Brapa_vs_Darmor_Bzh_V10_A,
                                  Darmor_Bzh_V10_A_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Darmor_Bzh_V10_C and Brapa
# Find the RBH between Darmor_Bzh_V10_C and Brapa
Brapa_vs_Darmor_Bzh_V10_C <- Brapa_vs_Darmor_Bzh_V10 %>%
  filter(grepl("^C",query.acc.ver))
Darmor_Bzh_V10_C_vs_Brapa <- Darmor_Bzh_V10_vs_Brapa %>%
  filter(grepl("^C",subject.acc.ver))

# Filter for best hits
Brapa_vs_Darmor_Bzh_V10_C <- Brapa_vs_Darmor_Bzh_V10_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Darmor_Bzh_V10_C_vs_Brapa <- Darmor_Bzh_V10_C_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Darmor_Bzh_V10_C <- inner_join(Brapa_vs_Darmor_Bzh_V10_C,
                                  Darmor_Bzh_V10_C_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Darmor_Bzh_V10_A and Boleracea
# Find the RBH between Darmor_Bzh_V10_A and Boleracea
Boleracea_vs_Darmor_Bzh_V10 <- read_table("BlastP/Boleracea_HDEM_vs_Darmor-Bzh_V10.out.gz", col_names = headers) %>% filter(!grepl("scaffold", query.acc.ver))
Boleracea_vs_Darmor_Bzh_V10_A <- Boleracea_vs_Darmor_Bzh_V10 %>%
  filter(grepl("^A",query.acc.ver))
Darmor_Bzh_V10_vs_Boleracea <- read_table("BlastP/Darmor-Bzh_V10_vs_Boleracea_HDEM.out.gz", col_names = headers) %>% filter(!grepl("scaffold", subject.acc.ver))
Darmor_Bzh_V10_A_vs_Boleracea <- Darmor_Bzh_V10_vs_Boleracea %>%
  filter(grepl("^A",subject.acc.ver))

# Filter for best hits
Boleracea_vs_Darmor_Bzh_V10_A <- Boleracea_vs_Darmor_Bzh_V10_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Darmor_Bzh_V10_A_vs_Boleracea <- Darmor_Bzh_V10_A_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_Darmor_Bzh_V10_A <- inner_join(Boleracea_vs_Darmor_Bzh_V10_A,
                                  Darmor_Bzh_V10_A_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Darmor_Bzh_V10_C and Boleracea
# Find the RBH between Darmor_Bzh_V10_C and Boleracea
Boleracea_vs_Darmor_Bzh_V10_C <- Boleracea_vs_Darmor_Bzh_V10 %>%
  filter(grepl("^C",query.acc.ver))
Darmor_Bzh_V10_C_vs_Boleracea <- Darmor_Bzh_V10_vs_Boleracea %>%
  filter(grepl("^C",subject.acc.ver))

# Filter for best hits
Boleracea_vs_Darmor_Bzh_V10_C <- Boleracea_vs_Darmor_Bzh_V10_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Darmor_Bzh_V10_C_vs_Boleracea <- Darmor_Bzh_V10_C_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_Darmor_Bzh_V10_C <- inner_join(Boleracea_vs_Darmor_Bzh_V10_C,
                                  Darmor_Bzh_V10_C_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Darmor_Bzh_V10_A and Darmor_Bzh_V10_C
# Find the RBH between Darmor_Bzh_V10_A and Darmor_Bzh_V10_C
Darmor_Bzh_V10_A_vs_Darmor_Bzh_V10_C <- read_table("BlastP/Darmor-Bzh_V10_A_vs_Darmor-Bzh_V10_C.out.gz", col_names = headers) %>% filter(!grepl("scaffold", query.acc.ver))
Darmor_Bzh_V10_C_vs_Darmor_Bzh_V10_A <- read_table("BlastP/Darmor-Bzh_V10_C_vs_Darmor-Bzh_V10_A.out.gz", col_names = headers) %>% filter(!grepl("scaffold", subject.acc.ver))

# Filter for best hits
Darmor_Bzh_V10_A_vs_Darmor_Bzh_V10_C <- Darmor_Bzh_V10_A_vs_Darmor_Bzh_V10_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Darmor_Bzh_V10_C_vs_Darmor_Bzh_V10_A <- Darmor_Bzh_V10_C_vs_Darmor_Bzh_V10_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Darmor_Bzh_V10_A_and_Darmor_Bzh_V10_C <- inner_join(Darmor_Bzh_V10_A_vs_Darmor_Bzh_V10_C,
                                  Darmor_Bzh_V10_C_vs_Darmor_Bzh_V10_A,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))


Darmor_Bzh_V10_Double_A <- HE_of_A(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_Darmor_Bzh_V10_A,
                         ArCn = Brapa_and_Darmor_Bzh_V10_C,
                         CoCn = Boleracea_and_Darmor_Bzh_V10_C,
                         AnCn = Darmor_Bzh_V10_A_and_Darmor_Bzh_V10_C,
                         name = "Darmor_Bzh_V10")


Darmor_Bzh_V10_Double_C <- HE_of_C(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_Darmor_Bzh_V10_A,
                         CoAn = Boleracea_and_Darmor_Bzh_V10_A,
                         CoCn = Boleracea_and_Darmor_Bzh_V10_C,
                         AnCn = Darmor_Bzh_V10_A_and_Darmor_Bzh_V10_C,
                         name = "Darmor_Bzh_V10")


Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Gene_Coverage.tsv.gz")
Darmor_Bzh_V10_Coverage <- Coverage_Readin("Combined_Coverages/Combined_Darmor-Bzh_V10_Gene_Coverage.tsv.gz")


DA_Pseudo_napus_Darmor_Bzh_V10 <- DA_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                      Pseudo_napus_Coverage,
                                      Brapa_and_Darmor_Bzh_V10_A,
                                      Darmor_Bzh_V10_Double_A,
                                      cutoff = 2.5)


DC_Pseudo_napus_Darmor_Bzh_V10 <- DC_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                                  Pseudo_napus_Coverage,
                                                  Boleracea_and_Darmor_Bzh_V10_C,
                                                  Darmor_Bzh_V10_Double_C,
                                                  cutoff = 2.5)


DA_Darmor_Bzh_V10 <- Double_Pseudo_napus_Formatter(Darmor_Bzh_V10_Coverage, Darmor_Bzh_V10_Double_A)
DC_Darmor_Bzh_V10 <- Double_Pseudo_napus_Formatter(Darmor_Bzh_V10_Coverage, Darmor_Bzh_V10_Double_C)

DA_Combined_Darmor_Bzh_V10 <- Coverage_Combiner(Brapa_and_Boleracea, DA_Pseudo_napus_Darmor_Bzh_V10, Brapa_and_Darmor_Bzh_V10_A, Boleracea_and_Darmor_Bzh_V10_C, DA_Darmor_Bzh_V10)
DC_Combined_Darmor_Bzh_V10 <- Coverage_Combiner(Brapa_and_Boleracea, DC_Pseudo_napus_Darmor_Bzh_V10, Brapa_and_Darmor_Bzh_V10_A, Boleracea_and_Darmor_Bzh_V10_C, DC_Darmor_Bzh_V10)

save(DA_Combined_Darmor_Bzh_V10, DC_Combined_Darmor_Bzh_V10, file = "Julin_Darmor-Bzh_V10_gene_doublings.RData")

DA_Darmor_Bzh_V10_Positions <- Position_Extract(DA_Combined_Darmor_Bzh_V10, Pseudo_napus_Coverage, Darmor_Bzh_V10_Coverage)
DC_Darmor_Bzh_V10_Positions <- Position_Extract(DC_Combined_Darmor_Bzh_V10, Pseudo_napus_Coverage, Darmor_Bzh_V10_Coverage)

save(DA_Darmor_Bzh_V10_Positions, DC_Darmor_Bzh_V10_Positions, file = "Julin_Darmor-Bzh_V10_Gene_Positions.RData")

rm(list = ls(pattern = "*Darmor*"))
```

## GanganF73
```{r}
# Repeat for DaAe_A and Brapa
# Find the RBH between DaAe_A and Brapa
Brapa_vs_GanganF73 <- read_table("BlastP/Brapa_Z1_V2_vs_GanganF73.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Brapa_vs_GanganF73_A <- Brapa_vs_GanganF73 %>%
  filter(grepl("BnaA",query.acc.ver))
GanganF73_vs_Brapa <- read_table("BlastP/GanganF73_vs_Brapa_Z1_V2.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
GanganF73_A_vs_Brapa <- GanganF73_vs_Brapa %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Brapa_vs_GanganF73_A <- Brapa_vs_GanganF73_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
GanganF73_A_vs_Brapa <- GanganF73_A_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_GanganF73_A <- inner_join(Brapa_vs_GanganF73_A,
                                  GanganF73_A_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for GanganF73_C and Brapa
# Find the RBH between GanganF73_C and Brapa
Brapa_vs_GanganF73_C <- Brapa_vs_GanganF73 %>%
  filter(grepl("BnaC",query.acc.ver))
GanganF73_C_vs_Brapa <- GanganF73_vs_Brapa %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Brapa_vs_GanganF73_C <- Brapa_vs_GanganF73_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
GanganF73_C_vs_Brapa <- GanganF73_C_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_GanganF73_C <- inner_join(Brapa_vs_GanganF73_C,
                                  GanganF73_C_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for GanganF73_A and Boleracea
# Find the RBH between GanganF73_A and Boleracea
Boleracea_vs_GanganF73 <- read_table("BlastP/Boleracea_HDEM_vs_GanganF73.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Boleracea_vs_GanganF73_A <- Boleracea_vs_GanganF73 %>%
  filter(grepl("BnaA",query.acc.ver))
GanganF73_vs_Boleracea <- read_table("BlastP/GanganF73_vs_Boleracea_HDEM.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
GanganF73_A_vs_Boleracea <- GanganF73_vs_Boleracea %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Boleracea_vs_GanganF73_A <- Boleracea_vs_GanganF73_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
GanganF73_A_vs_Boleracea <- GanganF73_A_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_GanganF73_A <- inner_join(Boleracea_vs_GanganF73_A,
                                  GanganF73_A_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for GanganF73_C and Boleracea
# Find the RBH between GanganF73_C and Boleracea
Boleracea_vs_GanganF73_C <- Boleracea_vs_GanganF73 %>%
  filter(grepl("BnaC",query.acc.ver))
GanganF73_C_vs_Boleracea <- GanganF73_vs_Boleracea %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Boleracea_vs_GanganF73_C <- Boleracea_vs_GanganF73_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
GanganF73_C_vs_Boleracea <- GanganF73_C_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_GanganF73_C <- inner_join(Boleracea_vs_GanganF73_C,
                                  GanganF73_C_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for GanganF73_A and GanganF73_C
# Find the RBH between GanganF73_A and GanganF73_C
GanganF73_A_vs_GanganF73_C <- read_table("BlastP/GanganF73_A_vs_GanganF73_C.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
GanganF73_C_vs_GanganF73_A <- read_table("BlastP/GanganF73_C_vs_GanganF73_A.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))

# Filter for best hits
GanganF73_A_vs_GanganF73_C <- GanganF73_A_vs_GanganF73_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
GanganF73_C_vs_GanganF73_A <- GanganF73_C_vs_GanganF73_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
GanganF73_A_and_GanganF73_C <- inner_join(GanganF73_A_vs_GanganF73_C,
                                  GanganF73_C_vs_GanganF73_A,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

GanganF73_Double_A <- HE_of_A(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_GanganF73_A,
                         ArCn = Brapa_and_GanganF73_C,
                         CoCn = Boleracea_and_GanganF73_C,
                         AnCn = GanganF73_A_and_GanganF73_C,
                         name = "GanganF73")


GanganF73_Double_C <- HE_of_C(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_GanganF73_A,
                         CoAn = Boleracea_and_GanganF73_A,
                         CoCn = Boleracea_and_GanganF73_C,
                         AnCn = GanganF73_A_and_GanganF73_C,
                         name = "GanganF73")

Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Gene_Coverage.tsv.gz")
GanganF73_Coverage <- Coverage_Readin("Combined_Coverages/Combined_GanganF73_Gene_Coverage.tsv.gz")

DA_Pseudo_napus_GanganF73 <- DA_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                      Pseudo_napus_Coverage,
                                      Brapa_and_GanganF73_A,
                                      GanganF73_Double_A,
                                      cutoff = 2.5)


DC_Pseudo_napus_GanganF73 <- DC_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                                  Pseudo_napus_Coverage,
                                                  Boleracea_and_GanganF73_C,
                                                  GanganF73_Double_C,
                                                  cutoff = 2.5)


DA_GanganF73 <- Double_Pseudo_napus_Formatter(GanganF73_Coverage, GanganF73_Double_A)
DC_GanganF73 <- Double_Pseudo_napus_Formatter(GanganF73_Coverage, GanganF73_Double_C)

DA_Combined_GanganF73 <- Coverage_Combiner(Brapa_and_Boleracea, DA_Pseudo_napus_GanganF73, Brapa_and_GanganF73_A, Boleracea_and_GanganF73_C, DA_GanganF73)
DC_Combined_GanganF73 <- Coverage_Combiner(Brapa_and_Boleracea, DC_Pseudo_napus_GanganF73, Brapa_and_GanganF73_A, Boleracea_and_GanganF73_C, DC_GanganF73)

save(DA_Combined_GanganF73, DC_Combined_GanganF73, file = "Julin_GanganF73_gene_doublings.RData")

DA_GanganF73_Positions <- Position_Extract(DA_Combined_GanganF73, Pseudo_napus_Coverage, GanganF73_Coverage)
DC_GanganF73_Positions <- Position_Extract(DC_Combined_GanganF73, Pseudo_napus_Coverage, GanganF73_Coverage)

save(DA_GanganF73_Positions, DC_GanganF73_Positions, file = "Julin_GanganF73_Gene_Positions.RData")
rm(list = ls(pattern = "*GanganF73*"))
```

## No2127
```{r}
# Repeat for DaAe_A and Brapa
# Find the RBH between DaAe_A and Brapa
Brapa_vs_No2127 <- read_table("BlastP/Brapa_Z1_V2_vs_No2127.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Brapa_vs_No2127_A <- Brapa_vs_No2127 %>%
  filter(grepl("BnaA",query.acc.ver))
No2127_vs_Brapa <- read_table("BlastP/No2127_vs_Brapa_Z1_V2.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
No2127_A_vs_Brapa <- No2127_vs_Brapa %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Brapa_vs_No2127_A <- Brapa_vs_No2127_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
No2127_A_vs_Brapa <- No2127_A_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_No2127_A <- inner_join(Brapa_vs_No2127_A,
                                  No2127_A_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for No2127_C and Brapa
# Find the RBH between No2127_C and Brapa
Brapa_vs_No2127_C <- Brapa_vs_No2127 %>%
  filter(grepl("BnaC",query.acc.ver))
No2127_C_vs_Brapa <- No2127_vs_Brapa %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Brapa_vs_No2127_C <- Brapa_vs_No2127_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
No2127_C_vs_Brapa <- No2127_C_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_No2127_C <- inner_join(Brapa_vs_No2127_C,
                                  No2127_C_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for No2127_A and Boleracea
# Find the RBH between No2127_A and Boleracea
Boleracea_vs_No2127 <- read_table("BlastP/Boleracea_HDEM_vs_No2127.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Boleracea_vs_No2127_A <- Boleracea_vs_No2127 %>%
  filter(grepl("BnaA",query.acc.ver))
No2127_vs_Boleracea <- read_table("BlastP/No2127_vs_Boleracea_HDEM.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
No2127_A_vs_Boleracea <- No2127_vs_Boleracea %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Boleracea_vs_No2127_A <- Boleracea_vs_No2127_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
No2127_A_vs_Boleracea <- No2127_A_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_No2127_A <- inner_join(Boleracea_vs_No2127_A,
                                  No2127_A_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for No2127_C and Boleracea
# Find the RBH between No2127_C and Boleracea
Boleracea_vs_No2127_C <- Boleracea_vs_No2127 %>%
  filter(grepl("BnaC",query.acc.ver))
No2127_C_vs_Boleracea <- No2127_vs_Boleracea %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Boleracea_vs_No2127_C <- Boleracea_vs_No2127_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
No2127_C_vs_Boleracea <- No2127_C_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_No2127_C <- inner_join(Boleracea_vs_No2127_C,
                                  No2127_C_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for No2127_A and No2127_C
# Find the RBH between No2127_A and No2127_C
No2127_A_vs_No2127_C <- read_table("BlastP/No2127_A_vs_No2127_C.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
No2127_C_vs_No2127_A <- read_table("BlastP/No2127_C_vs_No2127_A.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))

# Filter for best hits
No2127_A_vs_No2127_C <- No2127_A_vs_No2127_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
No2127_C_vs_No2127_A <- No2127_C_vs_No2127_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
No2127_A_and_No2127_C <- inner_join(No2127_A_vs_No2127_C,
                                  No2127_C_vs_No2127_A,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

No2127_Double_A <- HE_of_A(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_No2127_A,
                         ArCn = Brapa_and_No2127_C,
                         CoCn = Boleracea_and_No2127_C,
                         AnCn = No2127_A_and_No2127_C,
                         name = "No2127")


No2127_Double_C <- HE_of_C(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_No2127_A,
                         CoAn = Boleracea_and_No2127_A,
                         CoCn = Boleracea_and_No2127_C,
                         AnCn = No2127_A_and_No2127_C,
                         name = "No2127")

Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Gene_Coverage.tsv.gz")
No2127_Coverage <- Coverage_Readin("Combined_Coverages/Combined_No2127_Gene_Coverage.tsv.gz")

DA_Pseudo_napus_No2127 <- DA_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                      Pseudo_napus_Coverage,
                                      Brapa_and_No2127_A,
                                      No2127_Double_A,
                                      cutoff = 2.5)


DC_Pseudo_napus_No2127 <- DC_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                                  Pseudo_napus_Coverage,
                                                  Boleracea_and_No2127_C,
                                                  No2127_Double_C,
                                                  cutoff = 2.5)


DA_No2127 <- Double_Pseudo_napus_Formatter(No2127_Coverage, No2127_Double_A)
DC_No2127 <- Double_Pseudo_napus_Formatter(No2127_Coverage, No2127_Double_C)

DA_Combined_No2127 <- Coverage_Combiner(Brapa_and_Boleracea, DA_Pseudo_napus_No2127, Brapa_and_No2127_A, Boleracea_and_No2127_C, DA_No2127)
DC_Combined_No2127 <- Coverage_Combiner(Brapa_and_Boleracea, DC_Pseudo_napus_No2127, Brapa_and_No2127_A, Boleracea_and_No2127_C, DC_No2127)

save(DA_Combined_No2127, DC_Combined_No2127, file = "Julin_No2127_gene_doublings.RData")

DA_No2127_Positions <- Position_Extract(DA_Combined_No2127, Pseudo_napus_Coverage, No2127_Coverage)
DC_No2127_Positions <- Position_Extract(DC_Combined_No2127, Pseudo_napus_Coverage, No2127_Coverage)

save(DA_No2127_Positions, DC_No2127_Positions, file = "Julin_No2127_Gene_Positions.RData")

rm(list = ls(pattern = "*No2127*"))
```

## QuintaA
```{r}
# Repeat for DaAe_A and Brapa
# Find the RBH between DaAe_A and Brapa
Brapa_vs_QuintaA <- read_table("BlastP/Brapa_Z1_V2_vs_QuintaA.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Brapa_vs_QuintaA_A <- Brapa_vs_QuintaA %>%
  filter(grepl("BnaA",query.acc.ver))
QuintaA_vs_Brapa <- read_table("BlastP/QuintaA_vs_Brapa_Z1_V2.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
QuintaA_A_vs_Brapa <- QuintaA_vs_Brapa %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Brapa_vs_QuintaA_A <- Brapa_vs_QuintaA_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
QuintaA_A_vs_Brapa <- QuintaA_A_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_QuintaA_A <- inner_join(Brapa_vs_QuintaA_A,
                                  QuintaA_A_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for QuintaA_C and Brapa
# Find the RBH between QuintaA_C and Brapa
Brapa_vs_QuintaA_C <- Brapa_vs_QuintaA %>%
  filter(grepl("BnaC",query.acc.ver))
QuintaA_C_vs_Brapa <- QuintaA_vs_Brapa %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Brapa_vs_QuintaA_C <- Brapa_vs_QuintaA_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
QuintaA_C_vs_Brapa <- QuintaA_C_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_QuintaA_C <- inner_join(Brapa_vs_QuintaA_C,
                                  QuintaA_C_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for QuintaA_A and Boleracea
# Find the RBH between QuintaA_A and Boleracea
Boleracea_vs_QuintaA <- read_table("BlastP/Boleracea_HDEM_vs_QuintaA.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Boleracea_vs_QuintaA_A <- Boleracea_vs_QuintaA %>%
  filter(grepl("BnaA",query.acc.ver))
QuintaA_vs_Boleracea <- read_table("BlastP/QuintaA_vs_Boleracea_HDEM.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
QuintaA_A_vs_Boleracea <- QuintaA_vs_Boleracea %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Boleracea_vs_QuintaA_A <- Boleracea_vs_QuintaA_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
QuintaA_A_vs_Boleracea <- QuintaA_A_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_QuintaA_A <- inner_join(Boleracea_vs_QuintaA_A,
                                  QuintaA_A_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for QuintaA_C and Boleracea
# Find the RBH between QuintaA_C and Boleracea
Boleracea_vs_QuintaA_C <- Boleracea_vs_QuintaA %>%
  filter(grepl("BnaC",query.acc.ver))
QuintaA_C_vs_Boleracea <- QuintaA_vs_Boleracea %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Boleracea_vs_QuintaA_C <- Boleracea_vs_QuintaA_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
QuintaA_C_vs_Boleracea <- QuintaA_C_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_QuintaA_C <- inner_join(Boleracea_vs_QuintaA_C,
                                  QuintaA_C_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for QuintaA_A and QuintaA_C
# Find the RBH between QuintaA_A and QuintaA_C
QuintaA_A_vs_QuintaA_C <- read_table("BlastP/QuintaA_A_vs_QuintaA_C.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
QuintaA_C_vs_QuintaA_A <- read_table("BlastP/QuintaA_C_vs_QuintaA_A.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))

# Filter for best hits
QuintaA_A_vs_QuintaA_C <- QuintaA_A_vs_QuintaA_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
QuintaA_C_vs_QuintaA_A <- QuintaA_C_vs_QuintaA_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
QuintaA_A_and_QuintaA_C <- inner_join(QuintaA_A_vs_QuintaA_C,
                                  QuintaA_C_vs_QuintaA_A,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

QuintaA_Double_A <- HE_of_A(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_QuintaA_A,
                         ArCn = Brapa_and_QuintaA_C,
                         CoCn = Boleracea_and_QuintaA_C,
                         AnCn = QuintaA_A_and_QuintaA_C,
                         name = "QuintaA")


QuintaA_Double_C <- HE_of_C(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_QuintaA_A,
                         CoAn = Boleracea_and_QuintaA_A,
                         CoCn = Boleracea_and_QuintaA_C,
                         AnCn = QuintaA_A_and_QuintaA_C,
                         name = "QuintaA")

Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Gene_Coverage.tsv.gz")
QuintaA_Coverage <- Coverage_Readin("Combined_Coverages/Combined_QuintaA_Gene_Coverage.tsv.gz")

DA_Pseudo_napus_QuintaA <- DA_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                      Pseudo_napus_Coverage,
                                      Brapa_and_QuintaA_A,
                                      QuintaA_Double_A,
                                      cutoff = 2.5)


DC_Pseudo_napus_QuintaA <- DC_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                                  Pseudo_napus_Coverage,
                                                  Boleracea_and_QuintaA_C,
                                                  QuintaA_Double_C,
                                                  cutoff = 2.5)


DA_QuintaA <- Double_Pseudo_napus_Formatter(QuintaA_Coverage, QuintaA_Double_A)
DC_QuintaA <- Double_Pseudo_napus_Formatter(QuintaA_Coverage, QuintaA_Double_C)

DA_Combined_QuintaA <- Coverage_Combiner(Brapa_and_Boleracea, DA_Pseudo_napus_QuintaA, Brapa_and_QuintaA_A, Boleracea_and_QuintaA_C, DA_QuintaA)
DC_Combined_QuintaA <- Coverage_Combiner(Brapa_and_Boleracea, DC_Pseudo_napus_QuintaA, Brapa_and_QuintaA_A, Boleracea_and_QuintaA_C, DC_QuintaA)

save(DA_Combined_QuintaA, DC_Combined_QuintaA, file = "Julin_QuintaA_gene_doublings.RData")

DA_QuintaA_Positions <- Position_Extract(DA_Combined_QuintaA, Pseudo_napus_Coverage, QuintaA_Coverage)
DC_QuintaA_Positions <- Position_Extract(DC_Combined_QuintaA, Pseudo_napus_Coverage, QuintaA_Coverage)

save(DA_QuintaA_Positions, DC_QuintaA_Positions, file = "Julin_QuintaA_Gene_Positions.RData")

rm(list = ls(pattern = "*QuintaA*"))
```

## Shengli3
```{r}
# Repeat for DaAe_A and Brapa
# Find the RBH between DaAe_A and Brapa
Brapa_vs_Shengli3 <- read_table("BlastP/Brapa_Z1_V2_vs_Shengli3.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Brapa_vs_Shengli3_A <- Brapa_vs_Shengli3 %>%
  filter(grepl("BnaA",query.acc.ver))
Shengli3_vs_Brapa <- read_table("BlastP/Shengli3_vs_Brapa_Z1_V2.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
Shengli3_A_vs_Brapa <- Shengli3_vs_Brapa %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Brapa_vs_Shengli3_A <- Brapa_vs_Shengli3_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Shengli3_A_vs_Brapa <- Shengli3_A_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Shengli3_A <- inner_join(Brapa_vs_Shengli3_A,
                                  Shengli3_A_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Shengli3_C and Brapa
# Find the RBH between Shengli3_C and Brapa
Brapa_vs_Shengli3_C <- Brapa_vs_Shengli3 %>%
  filter(grepl("BnaC",query.acc.ver))
Shengli3_C_vs_Brapa <- Shengli3_vs_Brapa %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Brapa_vs_Shengli3_C <- Brapa_vs_Shengli3_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Shengli3_C_vs_Brapa <- Shengli3_C_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Shengli3_C <- inner_join(Brapa_vs_Shengli3_C,
                                  Shengli3_C_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Shengli3_A and Boleracea
# Find the RBH between Shengli3_A and Boleracea
Boleracea_vs_Shengli3 <- read_table("BlastP/Boleracea_HDEM_vs_Shengli3.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Boleracea_vs_Shengli3_A <- Boleracea_vs_Shengli3 %>%
  filter(grepl("BnaA",query.acc.ver))
Shengli3_vs_Boleracea <- read_table("BlastP/Shengli3_vs_Boleracea_HDEM.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
Shengli3_A_vs_Boleracea <- Shengli3_vs_Boleracea %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Boleracea_vs_Shengli3_A <- Boleracea_vs_Shengli3_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Shengli3_A_vs_Boleracea <- Shengli3_A_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_Shengli3_A <- inner_join(Boleracea_vs_Shengli3_A,
                                  Shengli3_A_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Shengli3_C and Boleracea
# Find the RBH between Shengli3_C and Boleracea
Boleracea_vs_Shengli3_C <- Boleracea_vs_Shengli3 %>%
  filter(grepl("BnaC",query.acc.ver))
Shengli3_C_vs_Boleracea <- Shengli3_vs_Boleracea %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Boleracea_vs_Shengli3_C <- Boleracea_vs_Shengli3_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Shengli3_C_vs_Boleracea <- Shengli3_C_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_Shengli3_C <- inner_join(Boleracea_vs_Shengli3_C,
                                  Shengli3_C_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Shengli3_A and Shengli3_C
# Find the RBH between Shengli3_A and Shengli3_C
Shengli3_A_vs_Shengli3_C <- read_table("BlastP/Shengli3_A_vs_Shengli3_C.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Shengli3_C_vs_Shengli3_A <- read_table("BlastP/Shengli3_C_vs_Shengli3_A.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))

# Filter for best hits
Shengli3_A_vs_Shengli3_C <- Shengli3_A_vs_Shengli3_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Shengli3_C_vs_Shengli3_A <- Shengli3_C_vs_Shengli3_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Shengli3_A_and_Shengli3_C <- inner_join(Shengli3_A_vs_Shengli3_C,
                                  Shengli3_C_vs_Shengli3_A,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

Shengli3_Double_A <- HE_of_A(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_Shengli3_A,
                         ArCn = Brapa_and_Shengli3_C,
                         CoCn = Boleracea_and_Shengli3_C,
                         AnCn = Shengli3_A_and_Shengli3_C,
                         name = "Shengli3")


Shengli3_Double_C <- HE_of_C(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_Shengli3_A,
                         CoAn = Boleracea_and_Shengli3_A,
                         CoCn = Boleracea_and_Shengli3_C,
                         AnCn = Shengli3_A_and_Shengli3_C,
                         name = "Shengli3")

Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Gene_Coverage.tsv.gz")
Shengli3_Coverage <- Coverage_Readin("Combined_Coverages/Combined_Shengli3_Gene_Coverage.tsv.gz")

DA_Pseudo_napus_Shengli3 <- DA_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                      Pseudo_napus_Coverage,
                                      Brapa_and_Shengli3_A,
                                      Shengli3_Double_A,
                                      cutoff = 2.5)


DC_Pseudo_napus_Shengli3 <- DC_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                                  Pseudo_napus_Coverage,
                                                  Boleracea_and_Shengli3_C,
                                                  Shengli3_Double_C,
                                                  cutoff = 2.5)


DA_Shengli3 <- Double_Pseudo_napus_Formatter(Shengli3_Coverage, Shengli3_Double_A)
DC_Shengli3 <- Double_Pseudo_napus_Formatter(Shengli3_Coverage, Shengli3_Double_C)

DA_Combined_Shengli3 <- Coverage_Combiner(Brapa_and_Boleracea, DA_Pseudo_napus_Shengli3, Brapa_and_Shengli3_A, Boleracea_and_Shengli3_C, DA_Shengli3)
DC_Combined_Shengli3 <- Coverage_Combiner(Brapa_and_Boleracea, DC_Pseudo_napus_Shengli3, Brapa_and_Shengli3_A, Boleracea_and_Shengli3_C, DC_Shengli3)

save(DA_Combined_Shengli3, DC_Combined_Shengli3, file = "Julin_Shengli3_gene_doublings.RData")

DA_Shengli3_Positions <- Position_Extract(DA_Combined_Shengli3, Pseudo_napus_Coverage, Shengli3_Coverage)
DC_Shengli3_Positions <- Position_Extract(DC_Combined_Shengli3, Pseudo_napus_Coverage, Shengli3_Coverage)

save(DA_Shengli3_Positions, DC_Shengli3_Positions, file = "Julin_Shengli3_Gene_Positions.RData")

rm(list = ls(pattern = "*Shengli3*"))
```

## Tapidor
```{r}
# Repeat for DaAe_A and Brapa
# Find the RBH between DaAe_A and Brapa
Brapa_vs_Tapidor <- read_table("BlastP/Brapa_Z1_V2_vs_Tapidor.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Brapa_vs_Tapidor_A <- Brapa_vs_Tapidor %>%
  filter(grepl("BnaA",query.acc.ver))
Tapidor_vs_Brapa <- read_table("BlastP/Tapidor_vs_Brapa_Z1_V2.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
Tapidor_A_vs_Brapa <- Tapidor_vs_Brapa %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Brapa_vs_Tapidor_A <- Brapa_vs_Tapidor_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Tapidor_A_vs_Brapa <- Tapidor_A_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Tapidor_A <- inner_join(Brapa_vs_Tapidor_A,
                                  Tapidor_A_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Tapidor_C and Brapa
# Find the RBH between Tapidor_C and Brapa
Brapa_vs_Tapidor_C <- Brapa_vs_Tapidor %>%
  filter(grepl("BnaC",query.acc.ver))
Tapidor_C_vs_Brapa <- Tapidor_vs_Brapa %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Brapa_vs_Tapidor_C <- Brapa_vs_Tapidor_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Tapidor_C_vs_Brapa <- Tapidor_C_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Tapidor_C <- inner_join(Brapa_vs_Tapidor_C,
                                  Tapidor_C_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Tapidor_A and Boleracea
# Find the RBH between Tapidor_A and Boleracea
Boleracea_vs_Tapidor <- read_table("BlastP/Boleracea_HDEM_vs_Tapidor.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Boleracea_vs_Tapidor_A <- Boleracea_vs_Tapidor %>%
  filter(grepl("BnaA",query.acc.ver))
Tapidor_vs_Boleracea <- read_table("BlastP/Tapidor_vs_Boleracea_HDEM.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
Tapidor_A_vs_Boleracea <- Tapidor_vs_Boleracea %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Boleracea_vs_Tapidor_A <- Boleracea_vs_Tapidor_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Tapidor_A_vs_Boleracea <- Tapidor_A_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_Tapidor_A <- inner_join(Boleracea_vs_Tapidor_A,
                                  Tapidor_A_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Tapidor_C and Boleracea
# Find the RBH between Tapidor_C and Boleracea
Boleracea_vs_Tapidor_C <- Boleracea_vs_Tapidor %>%
  filter(grepl("BnaC",query.acc.ver))
Tapidor_C_vs_Boleracea <- Tapidor_vs_Boleracea %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Boleracea_vs_Tapidor_C <- Boleracea_vs_Tapidor_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Tapidor_C_vs_Boleracea <- Tapidor_C_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_Tapidor_C <- inner_join(Boleracea_vs_Tapidor_C,
                                  Tapidor_C_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Tapidor_A and Tapidor_C
# Find the RBH between Tapidor_A and Tapidor_C
Tapidor_A_vs_Tapidor_C <- read_table("BlastP/Tapidor_A_vs_Tapidor_C.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Tapidor_C_vs_Tapidor_A <- read_table("BlastP/Tapidor_C_vs_Tapidor_A.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))

# Filter for best hits
Tapidor_A_vs_Tapidor_C <- Tapidor_A_vs_Tapidor_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Tapidor_C_vs_Tapidor_A <- Tapidor_C_vs_Tapidor_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Tapidor_A_and_Tapidor_C <- inner_join(Tapidor_A_vs_Tapidor_C,
                                  Tapidor_C_vs_Tapidor_A,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

Tapidor_Double_A <- HE_of_A(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_Tapidor_A,
                         ArCn = Brapa_and_Tapidor_C,
                         CoCn = Boleracea_and_Tapidor_C,
                         AnCn = Tapidor_A_and_Tapidor_C,
                         name = "Tapidor")


Tapidor_Double_C <- HE_of_C(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_Tapidor_A,
                         CoAn = Boleracea_and_Tapidor_A,
                         CoCn = Boleracea_and_Tapidor_C,
                         AnCn = Tapidor_A_and_Tapidor_C,
                         name = "Tapidor")

Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Gene_Coverage.tsv.gz")
Tapidor_Coverage <- Coverage_Readin("Combined_Coverages/Combined_Tapidor_Gene_Coverage.tsv.gz")

DA_Pseudo_napus_Tapidor <- DA_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                      Pseudo_napus_Coverage,
                                      Brapa_and_Tapidor_A,
                                      Tapidor_Double_A,
                                      cutoff = 2.5)


DC_Pseudo_napus_Tapidor <- DC_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                                  Pseudo_napus_Coverage,
                                                  Boleracea_and_Tapidor_C,
                                                  Tapidor_Double_C,
                                                  cutoff = 2.5)


DA_Tapidor <- Double_Pseudo_napus_Formatter(Tapidor_Coverage, Tapidor_Double_A)
DC_Tapidor <- Double_Pseudo_napus_Formatter(Tapidor_Coverage, Tapidor_Double_C)

DA_Combined_Tapidor <- Coverage_Combiner(Brapa_and_Boleracea, DA_Pseudo_napus_Tapidor, Brapa_and_Tapidor_A, Boleracea_and_Tapidor_C, DA_Tapidor)
DC_Combined_Tapidor <- Coverage_Combiner(Brapa_and_Boleracea, DC_Pseudo_napus_Tapidor, Brapa_and_Tapidor_A, Boleracea_and_Tapidor_C, DC_Tapidor)

save(DA_Combined_Tapidor, DC_Combined_Tapidor, file = "Julin_Tapidor_gene_doublings.RData")

DA_Tapidor_Positions <- Position_Extract(DA_Combined_Tapidor, Pseudo_napus_Coverage, Tapidor_Coverage)
DC_Tapidor_Positions <- Position_Extract(DC_Combined_Tapidor, Pseudo_napus_Coverage, Tapidor_Coverage)

save(DA_Tapidor_Positions, DC_Tapidor_Positions, file = "Julin_Tapidor_Gene_Positions.RData")


rm(list = ls(pattern = "*Tapidor*"))
```

## Westar
```{r}
# Repeat for DaAe_A and Brapa
# Find the RBH between DaAe_A and Brapa
Brapa_vs_Westar <- read_table("BlastP/Brapa_Z1_V2_vs_Westar.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Brapa_vs_Westar_A <- Brapa_vs_Westar %>%
  filter(grepl("BnaA",query.acc.ver))
Westar_vs_Brapa <- read_table("BlastP/Westar_vs_Brapa_Z1_V2.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
Westar_A_vs_Brapa <- Westar_vs_Brapa %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Brapa_vs_Westar_A <- Brapa_vs_Westar_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Westar_A_vs_Brapa <- Westar_A_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Westar_A <- inner_join(Brapa_vs_Westar_A,
                                  Westar_A_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Westar_C and Brapa
# Find the RBH between Westar_C and Brapa
Brapa_vs_Westar_C <- Brapa_vs_Westar %>%
  filter(grepl("BnaC",query.acc.ver))
Westar_C_vs_Brapa <- Westar_vs_Brapa %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Brapa_vs_Westar_C <- Brapa_vs_Westar_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Westar_C_vs_Brapa <- Westar_C_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Westar_C <- inner_join(Brapa_vs_Westar_C,
                                  Westar_C_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Westar_A and Boleracea
# Find the RBH between Westar_A and Boleracea
Boleracea_vs_Westar <- read_table("BlastP/Boleracea_HDEM_vs_Westar.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Boleracea_vs_Westar_A <- Boleracea_vs_Westar %>%
  filter(grepl("BnaA",query.acc.ver))
Westar_vs_Boleracea <- read_table("BlastP/Westar_vs_Boleracea_HDEM.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
Westar_A_vs_Boleracea <- Westar_vs_Boleracea %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Boleracea_vs_Westar_A <- Boleracea_vs_Westar_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Westar_A_vs_Boleracea <- Westar_A_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_Westar_A <- inner_join(Boleracea_vs_Westar_A,
                                  Westar_A_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Westar_C and Boleracea
# Find the RBH between Westar_C and Boleracea
Boleracea_vs_Westar_C <- Boleracea_vs_Westar %>%
  filter(grepl("BnaC",query.acc.ver))
Westar_C_vs_Boleracea <- Westar_vs_Boleracea %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Boleracea_vs_Westar_C <- Boleracea_vs_Westar_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Westar_C_vs_Boleracea <- Westar_C_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_Westar_C <- inner_join(Boleracea_vs_Westar_C,
                                  Westar_C_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Westar_A and Westar_C
# Find the RBH between Westar_A and Westar_C
Westar_A_vs_Westar_C <- read_table("BlastP/Westar_A_vs_Westar_C.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Westar_C_vs_Westar_A <- read_table("BlastP/Westar_C_vs_Westar_A.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))

# Filter for best hits
Westar_A_vs_Westar_C <- Westar_A_vs_Westar_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Westar_C_vs_Westar_A <- Westar_C_vs_Westar_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Westar_A_and_Westar_C <- inner_join(Westar_A_vs_Westar_C,
                                  Westar_C_vs_Westar_A,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

Westar_Double_A <- HE_of_A(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_Westar_A,
                         ArCn = Brapa_and_Westar_C,
                         CoCn = Boleracea_and_Westar_C,
                         AnCn = Westar_A_and_Westar_C,
                         name = "Westar")


Westar_Double_C <- HE_of_C(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_Westar_A,
                         CoAn = Boleracea_and_Westar_A,
                         CoCn = Boleracea_and_Westar_C,
                         AnCn = Westar_A_and_Westar_C,
                         name = "Westar")

Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Gene_Coverage.tsv.gz")
Westar_Coverage <- Coverage_Readin("Combined_Coverages/Combined_Westar_Gene_Coverage.tsv.gz")

DA_Pseudo_napus_Westar <- DA_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                      Pseudo_napus_Coverage,
                                      Brapa_and_Westar_A,
                                      Westar_Double_A,
                                      cutoff = 2.5)


DC_Pseudo_napus_Westar <- DC_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                                  Pseudo_napus_Coverage,
                                                  Boleracea_and_Westar_C,
                                                  Westar_Double_C,
                                                  cutoff = 2.5)


DA_Westar <- Double_Pseudo_napus_Formatter(Westar_Coverage, Westar_Double_A)
DC_Westar <- Double_Pseudo_napus_Formatter(Westar_Coverage, Westar_Double_C)

DA_Combined_Westar <- Coverage_Combiner(Brapa_and_Boleracea, DA_Pseudo_napus_Westar, Brapa_and_Westar_A, Boleracea_and_Westar_C, DA_Westar)
DC_Combined_Westar <- Coverage_Combiner(Brapa_and_Boleracea, DC_Pseudo_napus_Westar, Brapa_and_Westar_A, Boleracea_and_Westar_C, DC_Westar)

save(DA_Combined_Westar, DC_Combined_Westar, file = "Julin_Westar_gene_doublings.RData")

DA_Westar_Positions <- Position_Extract(DA_Combined_Westar, Pseudo_napus_Coverage, Westar_Coverage)
DC_Westar_Positions <- Position_Extract(DC_Combined_Westar, Pseudo_napus_Coverage, Westar_Coverage)

save(DA_Westar_Positions, DC_Westar_Positions, file = "Julin_Westar_Gene_Positions.RData")

rm(list = ls(pattern = "*Westar*"))
```

## Zheyou73
```{r}
# Repeat for DaAe_A and Brapa
# Find the RBH between DaAe_A and Brapa
Brapa_vs_Zheyou73 <- read_table("BlastP/Brapa_Z1_V2_vs_Zheyou73.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Brapa_vs_Zheyou73_A <- Brapa_vs_Zheyou73 %>%
  filter(grepl("BnaA",query.acc.ver))
Zheyou73_vs_Brapa <- read_table("BlastP/Zheyou73_vs_Brapa_Z1_V2.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
Zheyou73_A_vs_Brapa <- Zheyou73_vs_Brapa %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Brapa_vs_Zheyou73_A <- Brapa_vs_Zheyou73_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Zheyou73_A_vs_Brapa <- Zheyou73_A_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Zheyou73_A <- inner_join(Brapa_vs_Zheyou73_A,
                                  Zheyou73_A_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Zheyou73_C and Brapa
# Find the RBH between Zheyou73_C and Brapa
Brapa_vs_Zheyou73_C <- Brapa_vs_Zheyou73 %>%
  filter(grepl("BnaC",query.acc.ver))
Zheyou73_C_vs_Brapa <- Zheyou73_vs_Brapa %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Brapa_vs_Zheyou73_C <- Brapa_vs_Zheyou73_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Zheyou73_C_vs_Brapa <- Zheyou73_C_vs_Brapa %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Brapa_and_Zheyou73_C <- inner_join(Brapa_vs_Zheyou73_C,
                                  Zheyou73_C_vs_Brapa,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Zheyou73_A and Boleracea
# Find the RBH between Zheyou73_A and Boleracea
Boleracea_vs_Zheyou73 <- read_table("BlastP/Boleracea_HDEM_vs_Zheyou73.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Boleracea_vs_Zheyou73_A <- Boleracea_vs_Zheyou73 %>%
  filter(grepl("BnaA",query.acc.ver))
Zheyou73_vs_Boleracea <- read_table("BlastP/Zheyou73_vs_Boleracea_HDEM.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))
Zheyou73_A_vs_Boleracea <- Zheyou73_vs_Boleracea %>%
  filter(grepl("BnaA",subject.acc.ver))

# Filter for best hits
Boleracea_vs_Zheyou73_A <- Boleracea_vs_Zheyou73_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Zheyou73_A_vs_Boleracea <- Zheyou73_A_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_Zheyou73_A <- inner_join(Boleracea_vs_Zheyou73_A,
                                  Zheyou73_A_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Zheyou73_C and Boleracea
# Find the RBH between Zheyou73_C and Boleracea
Boleracea_vs_Zheyou73_C <- Boleracea_vs_Zheyou73 %>%
  filter(grepl("BnaC",query.acc.ver))
Zheyou73_C_vs_Boleracea <- Zheyou73_vs_Boleracea %>%
  filter(grepl("BnaC",subject.acc.ver))

# Filter for best hits
Boleracea_vs_Zheyou73_C <- Boleracea_vs_Zheyou73_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Zheyou73_C_vs_Boleracea <- Zheyou73_C_vs_Boleracea %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Boleracea_and_Zheyou73_C <- inner_join(Boleracea_vs_Zheyou73_C,
                                  Zheyou73_C_vs_Boleracea,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

# Repeat for Zheyou73_A and Zheyou73_C
# Find the RBH between Zheyou73_A and Zheyou73_C
Zheyou73_A_vs_Zheyou73_C <- read_table("BlastP/Zheyou73_A_vs_Zheyou73_C.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", query.acc.ver))
Zheyou73_C_vs_Zheyou73_A <- read_table("BlastP/Zheyou73_C_vs_Zheyou73_A.out.gz", col_names = headers) %>% filter(!grepl("Bnascaffold", subject.acc.ver))

# Filter for best hits
Zheyou73_A_vs_Zheyou73_C <- Zheyou73_A_vs_Zheyou73_C %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()
Zheyou73_C_vs_Zheyou73_A <- Zheyou73_C_vs_Zheyou73_A %>%
  group_by(query.acc.ver) %>%
  slice_max(order_by = bit.score, n = 1, with_ties = FALSE) %>%
  ungroup()

# Merge together
Zheyou73_A_and_Zheyou73_C <- inner_join(Zheyou73_A_vs_Zheyou73_C,
                                  Zheyou73_C_vs_Zheyou73_A,
                                  by = c("query.acc.ver" = "subject.acc.ver",
                                         "subject.acc.ver" = "query.acc.ver"))

Zheyou73_Double_A <- HE_of_A(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_Zheyou73_A,
                         ArCn = Brapa_and_Zheyou73_C,
                         CoCn = Boleracea_and_Zheyou73_C,
                         AnCn = Zheyou73_A_and_Zheyou73_C,
                         name = "Zheyou73")


Zheyou73_Double_C <- HE_of_C(ArCo = Brapa_and_Boleracea,
                         ArAn = Brapa_and_Zheyou73_A,
                         CoAn = Boleracea_and_Zheyou73_A,
                         CoCn = Boleracea_and_Zheyou73_C,
                         AnCn = Zheyou73_A_and_Zheyou73_C,
                         name = "Zheyou73")

Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Gene_Coverage.tsv.gz")
Zheyou73_Coverage <- Coverage_Readin("Combined_Coverages/Combined_Zheyou73_Gene_Coverage.tsv.gz")

DA_Pseudo_napus_Zheyou73 <- DA_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                      Pseudo_napus_Coverage,
                                      Brapa_and_Zheyou73_A,
                                      Zheyou73_Double_A,
                                      cutoff = 2.5)


DC_Pseudo_napus_Zheyou73 <- DC_Pseudo_napus_Formatter(Brapa_and_Boleracea,
                                                  Pseudo_napus_Coverage,
                                                  Boleracea_and_Zheyou73_C,
                                                  Zheyou73_Double_C,
                                                  cutoff = 2.5)


DA_Zheyou73 <- Double_Pseudo_napus_Formatter(Zheyou73_Coverage, Zheyou73_Double_A)
DC_Zheyou73 <- Double_Pseudo_napus_Formatter(Zheyou73_Coverage, Zheyou73_Double_C)

DA_Combined_Zheyou73 <- Coverage_Combiner(Brapa_and_Boleracea, DA_Pseudo_napus_Zheyou73, Brapa_and_Zheyou73_A, Boleracea_and_Zheyou73_C, DA_Zheyou73)
DC_Combined_Zheyou73 <- Coverage_Combiner(Brapa_and_Boleracea, DC_Pseudo_napus_Zheyou73, Brapa_and_Zheyou73_A, Boleracea_and_Zheyou73_C, DC_Zheyou73)

save(DA_Combined_Zheyou73, DC_Combined_Zheyou73, file = "Julin_Zheyou73_gene_doublings.RData")

DA_Zheyou73_Positions <- Position_Extract(DA_Combined_Zheyou73, Pseudo_napus_Coverage, Zheyou73_Coverage)
DC_Zheyou73_Positions <- Position_Extract(DC_Combined_Zheyou73, Pseudo_napus_Coverage, Zheyou73_Coverage)

save(DA_Zheyou73_Positions, DC_Zheyou73_Positions, file = "Julin_Zheyou73_Gene_Positions.RData")

rm(list = ls(pattern = "*Zheyou73*"))
```