---
title: "Sequence_Level_Homology_Julin_Ask"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
```

```{r}
## Read in the calculated coverages
coverage_columns <- c("Chromosome", "Start", "End", "Reads", "Covered", "Size", "Ratio")
Coverage_Readin <- function(filename){
  df <- read_delim(filename,
                   col_names = coverage_columns,
                   delim = "\t",
                   skip = 2) %>%
    mutate(Adjusted_Coverage = round(Reads / Size, 3),
           Position = floor((End + Start) / 2)) %>%
    mutate(Genome_Average = median(ifelse(Adjusted_Coverage>4,NA,Adjusted_Coverage), na.rm = T),
           Genome_SD = sd(Adjusted_Coverage, na.rm = T),
           Relative_Coverage = round(Adjusted_Coverage / Genome_Average, 3),
           Relative_Coverage = ifelse(Relative_Coverage >= 10, NA, Relative_Coverage),
           Std_Relative_Coverage = scale(Relative_Coverage))
  return(df)
}

Pseudo_napus_Coverage <- Coverage_Readin("Combined_Coverages/Combined_pseudo_napus_Coverage.tsv.gz")
DaAe_Coverage <- Coverage_Readin("Combined_Coverages/Combined_DaAe_Coverage.tsv.gz")
```

```{r}
## read in info on homoeoloous exchanges based on nucmer (Julin analysis)
load("Julin_HE_Comparisons.Rdata")
```


# Format the data a bit

```{r}
Pseudo_napus_Coverage <- Pseudo_napus_Coverage %>%
  mutate(Assembly = "Pseudo_napus",
         Chromosome = ifelse(grepl("C", Chromosome),
                             str_replace(Chromosome, "C", "chrC0"),
                             paste0("chr",Chromosome))
  )
DaAe_Coverage$Assembly <- "DaAe"
```

## Clustering a bit

```{r}
# Set the value of each point equal to median value in 20 neighbors will have to burn in 10
Clustered_Pseudo_napus_Coverage <- Pseudo_napus_Coverage %>%
  mutate(Consensus = NA)

for(i in 11:(nrow(Clustered_Pseudo_napus_Coverage)-10)){
  Neighbors <- Clustered_Pseudo_napus_Coverage$Relative_Coverage[(i-10):(i+10)]
  #winner <- as.numeric(names(table(neighbors))[table(neighbors)==max(table(neighbors))])
  Winner <- median(Neighbors, na.rm = T)
  if(is_empty(Winner)) Winner <- Clustered_Pseudo_napus_Coverage$Consensus[i-1]
  Clustered_Pseudo_napus_Coverage$Consensus[i] <- mean(Winner)
}

# Factor in Napus

Clustered_DaAe_Coverage <- DaAe_Coverage %>%
  mutate(Consensus = NA)

for(i in 11:(nrow(Clustered_DaAe_Coverage)-10)){
  Neighbors <- Clustered_DaAe_Coverage$Relative_Coverage[(i-10):(i+10)]
  #winner <- as.numeric(names(table(neighbors))[table(neighbors)==max(table(neighbors))])
  Winner <- median(Neighbors, na.rm = T)
  if(is_empty(Winner)) Winner <- Clustered_DaAe_Coverage$Consensus[i-1]
  Clustered_DaAe_Coverage$Consensus[i] <- mean(Winner)
}

Clustered_All <- rbind(Clustered_Pseudo_napus_Coverage, Clustered_DaAe_Coverage)
Clustered_All$Assembly <- factor(Clustered_All$Assembly, levels = c("Pseudo_napus", "DaAe"))
```

### Plots

```{r}
chroms <- c(paste0("chrA0",1:9), "chrA10", paste0("chrC0", 1:9))
for (i in chroms) {
  Clustered_All %>% filter(Chromosome == i) %>%
    mutate(Consensus = ifelse(Consensus >= 4.5, 4.5, Consensus)) %>%
    ggplot(aes(x = Position, y = Consensus)) +
    geom_point(aes(colour = cut(Consensus, c(-Inf, 0.6, 1.4, Inf)))) +
    scale_color_manual(name = "Consensus",
                     values = c("(-Inf,0.6]" = "red",
                                  "(0.6,1.4]" = "black",
                                  "(1.4, Inf]" = "blue"),
                     labels = c("<= 0.6", "0.6 < Consensus <= 1.4", "> 1.4")) +
    ggtitle(i) +
    geom_hline(yintercept = 1.5, color="black") +
    geom_hline(yintercept = 1, color="black") +
    geom_hline(yintercept = .5, color="black") +
    facet_wrap(~Assembly,
               ncol = 2,
               labeller = function (labels) {
                 labels <- lapply(labels, as.character)
                 labels[[1]] <- paste("Assembly:", labels[[1]])
                 a <-  do.call(paste,c(labels, list(sep = "\n")))
                 list(gsub("_","",a))
                 })
  #ggsave(paste0("Coverage_Plots_20221030/",i,".png"))
}
```


# Add in more napus genomes




Read in the coverages
```{r}
All_Coverage <- tibble(path=dir("Combined_Coverages", pattern = "Combine.*Coverage.tsv.gz", full.names = TRUE)) %>%
  filter(str_detect(path, "Boleracea|Brapa|Gene|V4", negate = TRUE)) # filter away files we do not want

All_Coverage <- All_Coverage %>%
  mutate(coverage=map(path, Coverage_Readin))
```

Format the data a bit and unnest

```{r}
#Create column for assembly names
All_Coverage <- All_Coverage %>%
  mutate(Assembly = str_remove_all(path, "(Combined_Coverages/Combined_|_Coverage.tsv.gz)"),
         Assembly = str_replace(Assembly, "pseudo", "Pseudo")) %>% # match rest of code
  select(Assembly, coverage) 

#Fix chromosome names
All_Coverage <- All_Coverage %>%
  mutate(coverage = map(coverage, mutate, Chromosome=str_extract(Chromosome, "[AC][0-9]{1,2}")),
         coverage = map(coverage, mutate, Chromosome=str_replace(Chromosome, "C([1-9])", "C0\\1")))

All_Coverage <- All_Coverage %>% unnest(coverage)
```

## Clustering a bit

```{r}
# Set the value of each point equal to most frequent rounded value in 20 neighbors will have to burn in 10

cluster_coverage <- function(relcov) {
  Consensus <- rep(NaN, length(relcov))
  for(i in 11:(length(relcov)-10)){
    Neighbors <- relcov[(i-10):(i+10)]
    Winner <- median(Neighbors, na.rm = T)
    if(is_empty(Winner)) Winner <- Consensus[i-1]
    Consensus[i] <- mean(Winner)
  }
  Consensus
}

Clustered_All <- All_Coverage %>%
  group_by(Assembly, Chromosome) %>%
  mutate(Consensus = cluster_coverage(Relative_Coverage))
```

# organize HE positions

Trying to compare John and Julin Lengths
```{r}
Clustered_All %>% group_by(Assembly, Chromosome) %>%
  filter(Assembly != "Pseudo_napus") %>%
  summarise(max(Position))
```


```{r}
levels(Clustered_All$Assembly)
unique(comparisons$assembly)
```


```{r}
HEpos <- comparisons %>% 
  filter(assembly %in% unique(Clustered_All$Assembly)) %>%
  select(Assembly=assembly, links) %>% 
  unnest(links) %>%
  mutate(Start=ref.start, 
         End=ref.end,
         Chromosome=str_replace(ref.name, "^.*([AC][0-9]{1,2})", "\\1"))

HEpos

HEpos %>% group_by(Assembly, Chromosome) %>%
  summarize(max(End))
```
# Load Position data
```{r}
load("Julin_DaAe_Gene_Positions.RData")
load("Julin_ZS11_Gene_Positions.RData")
load("Julin_Darmor-Bzh_V10_Gene_Positions.RData")
load("Julin_GanganF73_Gene_Positions.RData")
load("Julin_No2127_Gene_Positions.RData")
load("Julin_QuintaA_Gene_Positions.RData")
load("Julin_Shengli3_Gene_Positions.RData")
load("Julin_Tapidor_Gene_Positions.RData")
load("Julin_Westar_Gene_Positions.RData")
load("Julin_Zheyou73_Gene_Positions.RData")


name_fix <- function(positions, assembly_name = "", doubled = "A"){
  df <- positions %>%
    mutate(Chromosome = sub("chr", "", Chromosome),
           Chromosome = sub("scaffold", "", Chromosome),
           Chromosome = ifelse(Assembly == "Pseudo_napus",
                               str_replace(Chromosome,"C", "C0"),
                               Chromosome),
           Chromosome = sub(".*_", "", Chromosome),
           Chromosome = paste0("chr",Chromosome),
           Y = ifelse(grepl(doubled,Chromosome), 4.5, -.5),
           Shaped = ifelse(Y == 4.5, "<", ">"),
           Assembly = ifelse(Assembly == "Napus", assembly_name, Assembly)) %>%
    select(Chromosome, Position, Y, Assembly, Shaped)
  return(df)
}

DA_DaAe_Positions <- name_fix(DA_DaAe_Positions, "DaAe", "A")
DA_ZS11_Positions <- name_fix(DA_ZS11_Positions, "ZS11", "A")
DA_Darmor_Bzh_V10_Positions <- name_fix(DA_Darmor_Bzh_V10_Positions, "Darmor-Bzh_V10", "A")
DA_GanganF73_Positions <- name_fix(DA_GanganF73_Positions, "GanganF73", "A")
DA_No2127_Positions <- name_fix(DA_No2127_Positions, "No2127", "A")
DA_QuintaA_Positions <- name_fix(DA_QuintaA_Positions, "QuintaA", "A")
DA_Shengli3_Positions <- name_fix(DA_Shengli3_Positions, "Shengli3", "A")
DA_Tapidor_Positions <- name_fix(DA_Tapidor_Positions, "Tapidor", "A")
DA_Westar_Positions <- name_fix(DA_Westar_Positions, "Westar", "A")
DA_Zheyou73_Positions <- name_fix(DA_Zheyou73_Positions, "Zheyou73", "A")


DC_DaAe_Positions <- name_fix(DC_DaAe_Positions, "DaAe", "C")
DC_ZS11_Positions <- name_fix(DC_ZS11_Positions, "ZS11", "C")
DC_Darmor_Bzh_V10_Positions <- name_fix(DC_Darmor_Bzh_V10_Positions, "Darmor-Bzh_V10", "C")
DC_GanganF73_Positions <- name_fix(DC_GanganF73_Positions, "GanganF73", "C")
DC_No2127_Positions <- name_fix(DC_No2127_Positions, "No2127", "C")
DC_QuintaA_Positions <- name_fix(DC_QuintaA_Positions, "QuintaA", "C")
DC_Shengli3_Positions <- name_fix(DC_Shengli3_Positions, "Shengli3", "C")
DC_Tapidor_Positions <- name_fix(DC_Tapidor_Positions, "Tapidor", "C")
DC_Westar_Positions <- name_fix(DC_Westar_Positions, "Westar", "C")
DC_Zheyou73_Positions <- name_fix(DC_Zheyou73_Positions, "Zheyou73", "C")

All_Positions <- rbind(DA_DaAe_Positions, DA_ZS11_Positions, DA_Darmor_Bzh_V10_Positions, DA_GanganF73_Positions, DA_No2127_Positions, DA_QuintaA_Positions, DA_Shengli3_Positions, DA_Tapidor_Positions, DA_Westar_Positions, DA_Zheyou73_Positions, DC_DaAe_Positions, DC_ZS11_Positions, DC_Darmor_Bzh_V10_Positions, DC_GanganF73_Positions, DC_No2127_Positions, DC_QuintaA_Positions, DC_Shengli3_Positions, DC_Tapidor_Positions, DC_Westar_Positions, DC_Zheyou73_Positions
                       )
```


# Plot

```{r}
chroms <- unique(Clustered_All$Chromosome)
for (i in chroms) {
  Clustered_All %>% filter(Chromosome == i) %>%
    mutate(Consensus = ifelse(Consensus >= 4.5, 4.5, Consensus)) %>%
    mutate(Assembly = as.character(Assembly)) %>%
    ggplot(aes(x = Position, y = Consensus)) +
    geom_point(aes(colour = cut(Consensus, c(-Inf, 0.6, 1.4, Inf)))) +
    scale_color_manual(name = "Consensus",
                     values = c("(-Inf,0.6]" = "red",
                                  "(0.6,1.4]" = "black",
                                  "(1.4, Inf]" = "blue"),
                     labels = c("<= 0.6", "0.6 < Consensus <= 1.4", "> 1.4")) +
    geom_rect(data = {HEpos %>% filter(Chromosome == i)}, 
              aes(xmin=Start, xmax=End), 
              ymax=0, ymin=-1, fill="black", inherit.aes = FALSE) + 
    ggtitle(i) +
    geom_hline(yintercept = 1.5, color="blue") +
    geom_hline(yintercept = 1, color="black") +
    geom_hline(yintercept = .5, color="red") +
    facet_wrap(~Assembly,
               ncol = 2,
               labeller = function (labels) {
                 labels <- lapply(labels, as.character)
                 labels[[1]] <- paste("Assembly:", labels[[1]])
                 a <-  do.call(paste,c(labels, list(sep = "\n")))
                 list(gsub("_","",a))
                 }) +
    ylim(c(-0.5, 4.5))
  ggsave(paste0("Coverage_Plots_20221113/",i,".png"), width = 10, height = 10)
}

save(Clustered_All, All_Positions, file = "Julin_Coverages_and_Positions2_median.RData")
```

################
# Explore areas where gene and sequence level HE overlap

```{r}
library(tidyverse)
load("Julin_Coverages_and_Positions2.RData")
chroms <- c(paste0("chrA0",1:9), "chrA10", paste0("chrC0", 1:9))
for (i in chroms) {
  Clustered_All %>% filter(Chromosome == i) %>%
    mutate(Consensus = ifelse(Consensus >= 4,
                                          4,
                                          ifelse(Consensus <= -4,
                                                 -4,
                                                 Consensus))) %>%
    ggplot(aes(x = Position, y = Consensus)) +
    geom_point(aes(colour = cut(Consensus, c(-Inf, -1, 1, Inf)))) +
    scale_color_manual(name = "Consensus",
                     values = c("(-Inf,-1]" = "red",
                                  "(-1,1]" = "black",
                                  "(1, Inf]" = "blue"),
                     labels = c("<= -1", "-1 < Consensus <= 1", "> 1")) +
    ggtitle(i) +
    geom_hline(yintercept = 1, color="black") +
    geom_hline(yintercept = 0, color="black") +
    geom_hline(yintercept = -1, color="black") +
    facet_wrap(~Assembly,
               ncol = 2,
               labeller = function (labels) {
                 labels <- lapply(labels, as.character)
                 labels[[1]] <- paste("Assembly:", labels[[1]])
                 a <-  do.call(paste,c(labels, list(sep = "\n")))
                 list(gsub("_","",a))
                 }) #+
    #geom_text(data = All_Positions %>% filter(Chromosome == i),
               #aes(x = Position, y = Y, label = Shaped))
  ggsave(paste0("Coverage_Plots_20220929/",i,".png"), width = 10, height = 10)
}

```

# Get weird with it
```{r}
chroms <- c(paste0("chrA0", 1:9), "chrA10", paste0("chrC0", 1:9))
for (i in chroms) {
  Clustered_All %>% filter(Chromosome == i) %>%
    mutate(Consensus = ifelse(Consensus >= 4,
                              4,
                              ifelse(Consensus <= -4,-4,
                                     Consensus))) %>%
    ggplot(aes(x = Position, y = Consensus)) +
    geom_point(aes(colour = Assembly)) +
    ggtitle(i) +
    geom_hline(yintercept = 1, color = "black") +
    geom_hline(yintercept = 0, color = "black") +
    geom_hline(yintercept = -1, color = "black") +
    geom_text(
      data = All_Positions %>% filter(Chromosome == i),
      aes(
        x = Position,
        y = Y,
        label = Shaped,
        color = Assembly
      )
    )
  ggsave(
    paste0("Combined_Coverage_Plots_20220929/", i, ".png"),
    width = 10,
    height = 10
  )
}
```

```{r}
# Want regions greater than 1 std deviation in coverage
Filtered_Clustered_All<- Clustered_All #%>%
  #filter(Consensus <= -1 | Consensus >= 1)

# Want overlap of genes and sequence regions
DaAe_Seq <- Filtered_Clustered_All %>%
  filter(Assembly == "DaAe")
DaAe_Gene <- All_Positions %>%
  filter(Assembly == "DaAe")

Filtered_DaAe_Gene <- data.frame(matrix(ncol = ncol(DaAe_Gene), nrow = 0))
colnames(Filtered_DaAe_Gene) <- colnames(DaAe_Gene)
Filtered_DaAe_Seq <- data.frame(matrix(ncol = ncol(DaAe_Seq), nrow = 0))
colnames(Filtered_DaAe_Seq) <- colnames(DaAe_Seq)

# Get the genes in the regions
for(i in 1:nrow(DaAe_Gene)){
  chrom <- DaAe_Gene[[i,1]]
  pos <- DaAe_Gene[[i,2]]
  temp <- DaAe_Seq %>% filter(Chromosome == chrom)
  if(nrow(temp) == 0) next
  for(j in 1:nrow(temp)){
    if(temp[[j,2]] <= pos & temp[[j,3]] >= pos){
      Filtered_DaAe_Gene <- rbind(Filtered_DaAe_Gene, DaAe_Gene[i,])
      break
    }
  }
}

# Get the regions the genes are in
for(i in 1:nrow(DaAe_Seq)){
  chrom <- DaAe_Seq[[i,1]]
  start <- DaAe_Seq[[i,2]]
  end <- DaAe_Seq[[i,3]]
  temp <- Filtered_DaAe_Gene %>% filter(Chromosome == chrom)
  if(nrow(temp) == 0) next
  for(j in 1:nrow(temp)){
    if(temp[[j,2]] >= start & temp[[j,2]] <= end){
      Filtered_DaAe_Seq <- rbind(Filtered_DaAe_Seq, DaAe_Seq[i,])
      break
    }
  }
}

# Repeat for the other 3. This could be done a whole lot cleaner
# Want overlap of genes and sequence regions
ZS11_Seq <- Filtered_Clustered_All %>%
  filter(Assembly == "ZS11")
ZS11_Gene <- All_Positions %>%
  filter(Assembly == "ZS11")

Filtered_ZS11_Gene <- data.frame(matrix(ncol = ncol(ZS11_Gene), nrow = 0))
colnames(Filtered_ZS11_Gene) <- colnames(ZS11_Gene)
Filtered_ZS11_Seq <- data.frame(matrix(ncol = ncol(ZS11_Seq), nrow = 0))
colnames(Filtered_ZS11_Seq) <- colnames(ZS11_Seq)

# Get the genes in the regions
for(i in 1:nrow(ZS11_Gene)){
  chrom <- ZS11_Gene[[i,1]]
  pos <- ZS11_Gene[[i,2]]
  temp <- ZS11_Seq %>% filter(Chromosome == chrom)
  if(nrow(temp) == 0) next
  for(j in 1:nrow(temp)){
    if(temp[[j,2]] <= pos & temp[[j,3]] >= pos){
      Filtered_ZS11_Gene <- rbind(Filtered_ZS11_Gene, ZS11_Gene[i,])
      break
    }
  }
}

# Get the regions the genes are in
for(i in 1:nrow(ZS11_Seq)){
  chrom <- ZS11_Seq[[i,1]]
  start <- ZS11_Seq[[i,2]]
  end <- ZS11_Seq[[i,3]]
  temp <- Filtered_ZS11_Gene %>% filter(Chromosome == chrom)
  if(nrow(temp) == 0) next
  for(j in 1:nrow(temp)){
    if(temp[[j,2]] >= start & temp[[j,2]] <= end){
      Filtered_ZS11_Seq <- rbind(Filtered_ZS11_Seq, ZS11_Seq[i,])
      break
    }
  }
}

# Want overlap of genes and sequence regions
Darmor_Bzh_V10_Seq <- Filtered_Clustered_All %>%
  filter(Assembly == "Darmor-Bzh_V10")
Darmor_Bzh_V10_Gene <- All_Positions %>%
  filter(Assembly == "Darmor-Bzh_V10")

Filtered_Darmor_Bzh_V10_Gene <- data.frame(matrix(ncol = ncol(Darmor_Bzh_V10_Gene), nrow = 0))
colnames(Filtered_Darmor_Bzh_V10_Gene) <- colnames(Darmor_Bzh_V10_Gene)
Filtered_Darmor_Bzh_V10_Seq <- data.frame(matrix(ncol = ncol(Darmor_Bzh_V10_Seq), nrow = 0))
colnames(Filtered_Darmor_Bzh_V10_Seq) <- colnames(Darmor_Bzh_V10_Seq)

# Get the genes in the regions
for(i in 1:nrow(Darmor_Bzh_V10_Gene)){
  chrom <- Darmor_Bzh_V10_Gene[[i,1]]
  pos <- Darmor_Bzh_V10_Gene[[i,2]]
  temp <- Darmor_Bzh_V10_Seq %>% filter(Chromosome == chrom)
  if(nrow(temp) == 0) next
  for(j in 1:nrow(temp)){
    if(temp[[j,2]] <= pos & temp[[j,3]] >= pos){
      Filtered_Darmor_Bzh_V10_Gene <- rbind(Filtered_Darmor_Bzh_V10_Gene, Darmor_Bzh_V10_Gene[i,])
      break
    }
  }
}

# Get the regions the genes are in
for(i in 1:nrow(Darmor_Bzh_V10_Seq)){
  chrom <- Darmor_Bzh_V10_Seq[[i,1]]
  start <- Darmor_Bzh_V10_Seq[[i,2]]
  end <- Darmor_Bzh_V10_Seq[[i,3]]
  temp <- Filtered_Darmor_Bzh_V10_Gene %>% filter(Chromosome == chrom)
  if(nrow(temp) == 0) next
  for(j in 1:nrow(temp)){
    if(temp[[j,2]] >= start & temp[[j,2]] <= end){
      Filtered_Darmor_Bzh_V10_Seq <- rbind(Filtered_Darmor_Bzh_V10_Seq, Darmor_Bzh_V10_Seq[i,])
      break
    }
  }
}

# Want overlap of genes and sequence regions
Pseudo_napus_Seq <- Filtered_Clustered_All %>%
  filter(Assembly == "Pseudo_napus")
Pseudo_napus_Gene <- All_Positions %>%
  filter(Assembly == "Pseudo_napus")

Filtered_Pseudo_napus_Gene <- data.frame(matrix(ncol = ncol(Pseudo_napus_Gene), nrow = 0))
colnames(Filtered_Pseudo_napus_Gene) <- colnames(Pseudo_napus_Gene)
Filtered_Pseudo_napus_Seq <- data.frame(matrix(ncol = ncol(Pseudo_napus_Seq), nrow = 0))
colnames(Filtered_Pseudo_napus_Seq) <- colnames(Pseudo_napus_Seq)

# Get the genes in the regions
for(i in 1:nrow(Pseudo_napus_Gene)){
  chrom <- Pseudo_napus_Gene[[i,1]]
  pos <- Pseudo_napus_Gene[[i,2]]
  temp <- Pseudo_napus_Seq %>% filter(Chromosome == chrom)
  if(nrow(temp) == 0) next
  for(j in 1:nrow(temp)){
    if(temp[[j,2]] <= pos & temp[[j,3]] >= pos){
      Filtered_Pseudo_napus_Gene <- rbind(Filtered_Pseudo_napus_Gene, Pseudo_napus_Gene[i,])
      break
    }
  }
}

# Get the regions the genes are in
for(i in 1:nrow(Pseudo_napus_Seq)){
  chrom <- Pseudo_napus_Seq[[i,1]]
  start <- Pseudo_napus_Seq[[i,2]]
  end <- Pseudo_napus_Seq[[i,3]]
  temp <- Filtered_Pseudo_napus_Gene %>% filter(Chromosome == chrom)
  if(nrow(temp) == 0) next
  for(j in 1:nrow(temp)){
    if(temp[[j,2]] >= start & temp[[j,2]] <= end){
      Filtered_Pseudo_napus_Seq <- rbind(Filtered_Pseudo_napus_Seq, Pseudo_napus_Seq[i,])
      break
    }
  }
}

# Combine together
Combined_Gene <- rbind(Filtered_DaAe_Gene, Filtered_Darmor_Bzh_V10_Gene,
                       Filtered_ZS11_Gene, Filtered_Pseudo_napus_Gene)

Combined_Seq <- rbind(Filtered_DaAe_Seq, Filtered_Darmor_Bzh_V10_Seq,
                       Filtered_ZS11_Seq, Filtered_Pseudo_napus_Seq)

# Plot
chroms <- c(paste0("chrA0", 1:9), "chrA10", paste0("chrC0", 1:9))
for (i in chroms) {
  Combined_Seq %>% filter(Chromosome == i) %>%
    mutate(Consensus = ifelse(Consensus >= 4,
                              4,
                              ifelse(Consensus <= -4,-4,
                                     Consensus))) %>%
    ggplot(aes(x = Position, y = Consensus)) +
    geom_point(aes(colour = Assembly)) +
    ggtitle(i) +
    geom_hline(yintercept = 1, color = "black") +
    geom_hline(yintercept = 0, color = "black") +
    geom_hline(yintercept = -1, color = "black") +
    geom_text(
      data = Combined_Gene %>% filter(Chromosome == i),
      aes(
        x = Position,
        y = Y,
        label = Shaped,
        color = Assembly
      )
    )
  ggsave(
    paste0("Test_Combined_Coverage_Plots_20220929/", i, ".png"),
    width = 10,
    height = 10
  )
}
```

